---
title: "ST219_NDM_analysis"
author: "Tiffany Wan"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, results = FALSE, message = FALSE)
```




```{r, echo = FALSE}
# devtools::install_github('Snitkin-Lab-Umich/regentrans')
library(readxl)
library(data.table)
library(tidyverse)
# if(!require(devtools)) install.packages("devtools")
# devtools::install_github('Snitkin-Lab-Umich/regentrans')
#library(regentrans)
library(scatterpie)
library(ggthemes)
library(igraph)
library(ape)
library(phytools)
library(ggtree)
library(aplot)
library(graph4lg)
library(BSDA)
#install.packages("paletteer")
library(paletteer)
library(igraph)
library(scatterpie)
#install.packages("ggpubr")
library(ggpubr)
library(tidygraph)
library(ggraph)
library(janitor)
library(randomcoloR)
library(lubridate)
library(randomcoloR)
library(car)
library(cowplot)
library(scales)
library(directlabels)
library(ggrepel)
#if(!require(devtools)) install.packages("devtools")
#devtools::install_github("kassambara/rstatix")
library(rstatix)
library(pheatmap)
#install.packages("ggnetwork")
library(ggnetwork)
library(pheatmap)
#library(BactDating)
#install.packages("phylobase")
library(phylobase)
#install.packages("TransPhylo")
library(TransPhylo)
library(coda)
#install.packages("mcmcplots")
library(mcmcplots)
#library(seqinr)
library(ggnewscale)

```


```{r color palettes setup}
location_col <- c( "#5d8aa8", "#9966cc", "#a4c639", "#cd9575", "#915c83", "#8db600", "#195905", "#fbceb1", "#00ffff", "#7fffd4", "#e9d66b", "#89cff0", "#98777b", "#f4c2c2", "#de5d83", "#b5a642", "#ff55a3", "#cd7f32", "#f0dc82", "#cc5500", "#702963", "#5f9ea0", "#00bfff", "#ffa6c9", "#ace1af", "#ffa700", "#e4d00a", "#6f4e37", "#996666", "#008b8b", "#cd5b45", "#bdb76b", "#556b2f", "#872657", "#483d8b", "#704241", "#1560bd", "#614051", "#ccff00", "#cc6666", "#e48400", "#ccccff", "#008000", "#800000", "#915f6d", "red","black")
# 
# location_col <- c("#ffc0cb", "#6495ed", "#dda0dd", "#afeeee", "#ee82ee", "#98fb98", "#7fffd4", "#ffdead", "#ff69b4", "#2f4f4f", "#556b2f", "#8b4513", "#2e8b57", "#191970", "#8b0000", "#808000", "#008000", "#008080", "#b8860b", "#bdb76b", "#4682b4", "#d2691e", "#9acd32", "#00008b", "#32cd32", "#7f007f", "#8fbc8f", "#b03060", "#ffa500", "#ffa500", "#ffd700", "#ffff00", "#0000cd", "#00ff00", "#9400d3", "#00fa9a", "#dc143c", "#00ffff", "#00bfff", "#f4a460", "#9370db", "#f08080", "#adff2f", "#ff6347", "#b0c4de",  "red","black")

all_HCF_v <- c("Villa at Rose City","Ascension St. John","Beaumont Grosse Pointe","Mission Point Nursing & Physical Rehab Center of Detroit","DMC Detroit Receiving Hospital","Vibra DMC Campus","Mid-Michigan Hospital West Branch","DMC Harper University Hospital","Heartland Healthcare Center Grosse Pointe Woods","DMC Sinai Grace Hospital","DMC Rehabilitation Institute of Michigan","Riverview Health & Rehab Center Jefferson","Ascension Macomb Oakland - Warren Campus","Henry Ford Hospital","Father Murray, A Villa Center","McLaren Port Huron","Regency on the Lake","Lake Huron Medical Center","Ascension Macomb Oakland - Madison Heights Campus","Ascension Providence Novi","Hartford Nursing & Rehab","Select Specialty Hospital Northwest Detroit", "Select Specialty Hospital Grosse Pointe","Omni Continuing Care","Beaumont Royal Oak","Henry Ford Macomb Hospital", "Ascension Providence Park Southfield","Select Specialty Hospital Downriver","Beaumont Dearborn", "Advantage Living Center Samaritan", "Henry Ford Wyandotte Hospital", "St. Joseph's Healthcare Center","Karmanos Cancer Center", "Ambassador, A Villa Place","Spectrum Butterworth Hospital","Wellbridge Rochester Hills", "Divinity Home Care", "Regency at Chene", "Hamilton Nursing & Rehabilitation", "John D. Dingell Department of Veterans Affairs Medical Center","Riverview Health & Rehab Center North","Vibra Taylor", "Beaumont Taylor", "Michigan Medicine", "Regency at Bluffs Park")
#unique(patient_loc_ID_long1$HCF)
location_col <- setNames(location_col , c(all_HCF_v ,"Surveillance","Clinical"))

location_col_df <- data.frame(HCF = all_HCF_v,
                              color = location_col[1:(length(location_col)-2)] )

```

```{r}
location_col <- c( "#800000",
                   "#de5d83",
                   "#872657",
                   "#cc6666",
                   "#cd5b45",
                   "#cc5500",
                   "#ffa700",
                   "#e48400",
                   "#cd7f32",
                   "#ffd700",
                   "#483d8b",
                   "#e5e500",
                   "#f4c2c2",
                   "#ffa6c9",
                   "#00ffff",
                   "#98777b",
                   "#996666",
                   "#fbceb1",
                   "#6f4e37",
                   "#808000",
                   "#702963",
                   "#00bfff",
                   "#915c83",
                   "#ff55a3",
                   "#704241",
                   "#bdb76b",
                   "#556b2f",
                   "#5f9ea0",
                   "#008000",
                   "#ccff00",
                   "#e0ffff",
                   "#ace1af",
                   "#7fffd4",
                   "#ccccff",
                   "#614051",
                   "#8db600",
                   "#a4c639",
                   "#cd9575",
                   "#008b8b",
                   "#89cff0",
                   "#195905",
                   "#004953",
                   "#9966cc",
                   "#e9d66b",
                   "#5d8aa8",
                   "#1560bd",
                   "#915f6d",
                    "red",
                    "black")

all_HCF_v <- c("Ascension Macomb Oakland - Madison Heights Campus",
               "Ascension Macomb Oakland - Warren Campus",
               "Ascension Providence Novi",
               "Ascension Providence Park Southfield",
               "Ascension St. John",
               "Beaumont Dearborn",
               "Beaumont Grosse Pointe",
               "Beaumont Royal Oak",
               "Beaumont Taylor",
               "DMC Detroit Receiving Hospital",
               "DMC Harper University Hospital",
               "DMC Sinai Grace Hospital",
               "Henry Ford Hospital",
               "Henry Ford Macomb Hospital",
               "Henry Ford Wyandotte Hospital",
               "Lake Huron Medical Center",
               "McLaren Port Huron",
               "Michigan Medicine",
               "Mid-Michigan Hospital West Branch",
               "Spectrum Butterworth Hospital",
               "John D. Dingell Department of Veterans Affairs Medical Center",
               "Karmanos Cancer Center",
               "Divinity Home Care",
               "DMC Rehabilitation Institute of Michigan",
               "Select Specialty Hospital Ann Arbor",
               "Select Specialty Hospital Downriver",
               "Select Specialty Hospital Grosse Pointe",
               "Select Specialty Hospital Northwest Detroit",
               "Vibra DMC Campus",
               "Vibra Taylor",
               "Advantage Living Center Northwest",
               "Advantage Living Center Samaritan",
               "Ambassador, A Villa Place",
               "Father Murray, A Villa Center",
               "Hamilton Nursing & Rehabilitation",
               "Hartford Nursing & Rehab",
               "Heartland Healthcare Center Grosse Pointe Woods",
               "Mission Point Nursing & Physical Rehab Center of Detroit",
               "Omni Continuing Care",
               "Regency at Bluffs Park",
               "Regency at Chene",
               "Regency on the Lake",
               "Riverview Health & Rehab Center Jefferson",
               "Riverview Health & Rehab Center North",
               "St. Joseph's Healthcare Center",
               "Villa at Rose City",
               "Wellbridge Rochester Hills")

# location_col <- c("#ffc0cb", "#6495ed", "#dda0dd", "#afeeee", "#ee82ee", "#98fb98", "#7fffd4", "#ffdead", "#ff69b4", "#2f4f4f", "#556b2f", "#8b4513", "#2e8b57", "#191970", "#8b0000", "#808000", "#008000", "#008080", "#b8860b", "#bdb76b", "#4682b4", "#d2691e", "#9acd32", "#00008b", "#32cd32", "#7f007f", "#8fbc8f", "#b03060", "#ffa500", "#ffa500", "#ffd700", "#ffff00", "#0000cd", "#00ff00", "#9400d3", "#00fa9a", "#dc143c", "#00ffff", "#00bfff", "#f4a460", "#9370db", "#f08080", "#adff2f", "#ff6347", "#b0c4de",  "red","black")
#unique(patient_loc_ID_long1$HCF)
location_col <- setNames(location_col , c(all_HCF_v ,"Surveillance","Clinical"))
location_col_df <- data.frame(HCF = all_HCF_v,
                              color = location_col[1:(length(location_col)-2)] )
```


```{r}
# read in metadata of HCF type information
HCF_info <- fread("/Users/tifwan/RA/MDHSS/data/HCF_info_coord.csv") %>%
  select(-V1)
HCF_info$HCF_CCN <- gsub("A","",HCF_info$HCF_CCN)

# read in masked variant calling output
dna_aln <- read.dna("/Users/tifwan/Desktop/flux_mount_gl/Project_MDHHS_genomics/old_Sequence_data/ST219_Public_genomes/2022_08_24_MDHHS_Public_variant_calling/2022_08_25_16_34_37_core_results/gubbins/2022_08_25_16_34_37_SAMN26729713_genome_aln_w_alt_allele_unmapped.filtered_polymorphic_sites.fasta", format = "fasta", as.matrix=TRUE,as.character = "TRUE")

# read in patient trace data
setwd("~/RA/MDHSS/R_script")
patient_loc <- read.csv("../data/2022_08_25_location_time.csv")
patient_ID <- read_xlsx("/Users/tifwan/Dropbox (University of Michigan)/Michigan_CRE_Genomics/NDM Kpneumoniae ST219 7-11-22_update 10-4.xlsx",sheet = 2)
MDHHS_list <- fread("../data/MI_isolate_accn.tsv",)
iqtree <- read.tree("/Users/tifwan/Desktop/flux_mount_gl/Project_MDHHS_genomics/old_Sequence_data/ST219_Public_genomes/2022_08_24_MDHHS_Public_variant_calling/2022_08_25_16_34_37_core_results/gubbins/iqtree_results/2022_08_25_16_34_37_SAMN26729713_genome_aln_w_alt_allele_unmapped.treefile")

# key matching Isolate_ID, NCBI_SAMN
key <- fread("/Users/tifwan/Dropbox (University of Michigan)/Michigan_CRE_Genomics/Genome meta-data/st219_lookup_table.txt")

isolate_pt <- read_xlsx("~/RA/MDHSS/data/isolate_ID_PatientID_match.xlsx")
```


```{r}
# rename HCF_name variables to match the patient HCF transfer network
HCF_info$HCF_name<- gsub("Advantage Living Center Samaritan.*","Advantage Living Center Samaritan",HCF_info$HCF_name)
HCF_info$HCF_name<- gsub("Advantage Living Center Northwest.*","Advantage Living Center Northwest",HCF_info$HCF_name)
HCF_info$HCF_name<- gsub("Advantage Living Center Northwest.*","Advantage Living Center Northwest",HCF_info$HCF_name)
HCF_info$HCF_name<- gsub("Heartland Healthcare Center Grosse Pointe Woods.*","Heartland Healthcare Center Grosse Pointe Woods",HCF_info$HCF_name)
HCF_info$HCF_name<- gsub("Divinity Home Care.*","Divinity Home Care",HCF_info$HCF_name)
HCF_info$HCF_name<- gsub("DMC Rehabillitation Institute of Michigan","DMC Rehabilitation Institute of Michigan",HCF_info$HCF_name)
HCF_info$HCF_name<- gsub("Ascension Providence Southfield","Ascension Providence Park Southfield",HCF_info$HCF_name)
HCF_info$HCF_name<- gsub("Ascension Providence Southfield","Ascension Providence Park Southfield",HCF_info$HCF_name)

HCF_info$HCF_name <- trimws(HCF_info$HCF_name) # trim white space

MDHHS_CCN <- HCF_info %>%
  select(HCF_name, HCF_CCN)

HCF_coord <- HCF_info %>%
  select(HCF_name, lon, lat)

# subset HCF w/ coordinates
HCF_type <- HCF_info %>%
  select(HCF_name, HCF_type)

```


```{r }
# fix the isolate ID and corresponding NCBI
key[Isolate_ID=="20MP010425","NCBI_SAMN"] <- "20MP010425-WI-M3453-201110"
key[Isolate_ID=="20MP010444","NCBI_SAMN"] <- "20MP010444-WI-M5194-201116_S17"

# get Isolate_ID and Patient_ID
key2 <- patient_ID %>%
  select(Isolate_ID, Patient_ID)
key3 <- merge(key, key2, by = "Isolate_ID")
patient_ID_key <- merge(patient_ID, key,by = "Isolate_ID")
patient_type <- patient_ID_key %>%
  select(NCBI_SAMN,`Isolate Type`)


# renaming observations loc_date
patient_loc <- apply(patient_loc,2,function(x){gsub('Henry Ford Macomb Hospital Hospital','Henry Ford Macomb Hospital',x)})
patient_loc <- apply(patient_loc,2,function(x){gsub('DMC Rehabillitation Institute of Michigan','DMC Rehabilitation Institute of Michigan',x)})
patient_loc <- apply(patient_loc,2,function(x){gsub('Regency at Bluffs','Regency at Bluffs Park',x)})
patient_loc <- apply(patient_loc,2,function(x){gsub('Regency at Bluffs Park Park','Regency at Bluffs Park',x)})
patient_loc <- apply(patient_loc,2,function(x){gsub('Ascension Providence Southfield','Ascension Providence Park Southfield',x)})
patient_loc <- as.data.frame(patient_loc)


# merge key with location data
patient_loc_key <- merge(key3, patient_loc,  by = "Patient_ID", all = TRUE) 
patient_loc_key <- patient_loc_key %>%
  mutate(Isolate_ID = Isolate_ID.x) %>%
  select(-c(Isolate_ID.x, Isolate_ID.y))


# merge location and isolate collection dates
patient_loc_key_ID <- merge(patient_ID, patient_loc_key, by = "Isolate_ID", all = TRUE)
patient_loc_key_ID <- as.data.table(patient_loc_key_ID) # change df to data.table before tranposing to long 

patient_loc_key_ID <- patient_loc_key_ID %>%
  mutate(Patient_ID = Patient_ID.x) %>%
  select(-c(Patient_ID.x,Patient_ID.y))

patient_loc_key_ID <- arrange(patient_loc_key_ID, Specimen_collection_date)


# make sure the name of the tree label.tips matches the list 
MDHHS_list1 <- MDHHS_list
MDHHS_list1[MDHHS_list1$x=="20MP010425-WI-M3453-201110_S1",2] <- "20MP010425-WI-M3453-201110"

##
```

```{r}
## Isolate type
print("Number of Patients")
print("Isolate type")
table(patient_loc_key_ID$`Isolate Type`)
```


```{r }
# only keep MDHHS samples and the outgroup
dna_aln2 <- as.data.frame(dna_aln) 
dna_aln2 <- filter(dna_aln2, rownames(dna_aln2)%in%MDHHS_list1$x|rownames(dna_aln2)=='SAMEA3531785')
dna_aln <- as.matrix(dna_aln2)

# get the index of the outgroup
out_group <- which(row.names(dna_aln)=="SAMEA3531785") 
class(dna_aln)

# filter out outgroup
dna_aln_no_outgroup <- dna_aln[-which(row.names(dna_aln)=="SAMEA3531785"),] 
class(dna_aln_no_outgroup)

# return named vector with TRUE/FALSE (whether that position has more than one allele)
multi_allele <- apply(dna_aln_no_outgroup,
                      2,
                      FUN = function(aln_pos)
                      {
                        
                        pos_alleles <- unique(aln_pos)
                       # print(pos_alleles)
                        
                        dna_aln_no_outgroup <- setdiff(pos_alleles, c('-', 'n')) # setdiff(a,b): find all values in a that do not occur in b, aka find all unique values besides'-','n' 
                        
                        return(length(dna_aln_no_outgroup) > 1) # get the positions where there were more than one alleles
                        
                      })

dna_aln3 <- dna_aln[,which(multi_allele==TRUE)] # subset dna_aln to postions with more than one allele 
dna_aln <- dna_aln3
#t <- combn(nrow(dna_aln),2)

# function from prof: to create shared variants matrix
dna_shared_mat = matrix(0, nrow = nrow(dna_aln), ncol = nrow(dna_aln), dimnames = list(row.names(dna_aln), row.names(dna_aln)))
dna_shared_vec <- apply(combn(nrow(dna_aln), 2), 2, FUN = function(x){ # create dataset with all combinations of pair-wise comparison of 
  sum(dna_aln[x[1],] != dna_aln[out_group, ] &  # from each pair-wise comparision, if the alleles from the first one in the pair was different from the outgroup allele (gets boolean), then add up those that were true 
        dna_aln[x[1], ] == dna_aln[x[2], ] & # see if alleles between the pairwise comparison (boolean), then add up those that were true
        dna_aln[x[1], ] != "-" & # exclude those that were "-"
        dna_aln[x[1], ] != 'n' & # exclude those that were "n"
        dna_aln[out_group, ] != "-" & # exclude those that were "-"
        dna_aln[out_group, ] != 'n') # exclude those that were "n"
})
dna_shared_mat[t(combn(nrow(dna_aln), 2))] = dna_shared_vec # upper triangle
dna_shared_mat[t(combn(nrow(dna_aln), 2))[,2:1]] = dna_shared_vec # lower triangle
```



```{r long form patient transfer data}
# Reformat MDHHS transfer data to long
# tranpose from wide to long format
patient_loc_key_ID_long <- melt(patient_loc_key_ID, id.vars = c("NCBI_SAMN"),
                                measure.vars = list(HCF = c(paste0("HCF_exposure_",seq(1,7,1))),
                                                    AdmitDate = c(paste0("HCF_",seq(1,7,1),"_admit_dt")),
                                                    DischargeDate = c(paste0("HCF_",seq(1,7,1),"_discharge_dt")),
                                                    CollectionDate = "Specimen_collection_date",
                                                    Patient_ID = rep(c("Patient_ID"),7),
                                                    Isolate_ID = rep(c("Isolate_ID"),7),
                                                    Isolate_type = rep(c("Isolate Type"),7)))

patient_loc_key_ID_long$HCF <- trimws(patient_loc_key_ID_long$HCF)
patient_loc_ID_long <- patient_loc_key_ID_long
# change variable formats to date
patient_loc_ID_long$AdmitDate <- as.Date(patient_loc_ID_long$AdmitDate, format = "%m/%d/%y")
patient_loc_ID_long$DischargeDate <- as.Date(patient_loc_ID_long$DischargeDate, format = "%m/%d/%y")
patient_loc_ID_long$CollectionDate <- as.Date(patient_loc_ID_long$CollectionDate, format = "%Y-%m-%d")

HCF_location <- unique(patient_loc_ID_long$HCF)
# filter out missing location/date data and possibly incorrect dates
```


```{r data cleaning of long form}
patient_loc_ID_long1 <- patient_loc_ID_long %>%
  filter(HCF!="") %>%
  filter(NCBI_SAMN!="SAMN24979401")

patient_loc_ID_long1 <- patient_loc_ID_long1 %>%
  arrange(NCBI_SAMN, variable) %>%
  mutate(DischargeDate=as.Date(ifelse(is.na(DischargeDate), lag(AdmitDate),DischargeDate), origin = "1970-01-01"),
         AdmitDate=as.Date(ifelse(is.na(AdmitDate), DischargeDate,AdmitDate),origin ="1970-01-01"))

#patient_loc_ID_long1 <- patient_loc_ID_long1 %>%
#  filter(!is.na(AdmitDate)&!is.na(DischargeDate))

## Find abnormal/missing AdmitDates dates
abnormal_Admit_v <- which(patient_loc_ID_long1$AdmitDate<as.Date("2019-01-01")|is.na(patient_loc_ID_long1$AdmitDate)) # get all transfer admit dates that don't make sense
patient_loc_ID_long1$NCBI_SAMN[abnormal_Admit_v ]
weird_admit_df <- patient_loc_ID_long1[patient_loc_ID_long1$NCBI_SAMN%in%patient_loc_ID_long1$NCBI_SAMN[abnormal_Admit_v ],]

## Find abnormal Discharge dates
abnormal_Discharge_v <-which(patient_loc_ID_long1$DischargeDate<as.Date("2019-01-01")|is.na(patient_loc_ID_long1$DischargeDate)) # get all transfer discharge dates that don't make sense
patient_loc_ID_long1$NCBI_SAMN[abnormal_Discharge_v]
weird_discharge_df <- patient_loc_ID_long1[patient_loc_ID_long1$NCBI_SAMN%in%patient_loc_ID_long1$NCBI_SAMN[abnormal_Discharge_v],]

weird_df <- rbind(weird_admit_df, weird_discharge_df) %>%
  distinct() 

# impute the suspicious dates with probable dates
weird_df <- weird_df %>%
  arrange(NCBI_SAMN, desc(variable)) %>%
  group_by(NCBI_SAMN) %>%
  mutate(to_change = ifelse(AdmitDate<as.Date("2019-01-01")|DischargeDate<as.Date("2019-01-01")|is.na(AdmitDate)|is.na(DischargeDate),1,0),
         AdmitDate_new = ifelse(AdmitDate<as.Date("2019-01-01"), lag(DischargeDate), AdmitDate),
         AdmitDate_new = case_when(is.na(AdmitDate_new)&is.na(DischargeDate)~lead(AdmitDate-1),
                                   is.na(AdmitDate_new)~DischargeDate-1,
                                   TRUE ~  AdmitDate),
         DischargeDate_new = ifelse(DischargeDate<as.Date("2019-01-01")|is.na(DischargeDate), lead(AdmitDate),DischargeDate),
         AdmitDate = AdmitDate_new,
         DischargeDate = DischargeDate_new) %>%
  select(-c(AdmitDate_new, DischargeDate_new)) %>%
  filter(to_change==1) %>% # only keep isolates that I want to change
  select(-to_change)

weird_df$AdmitDate <- as.Date(weird_df$AdmitDate,origin ="1970-01-01")  
weird_df$DischargeDate <- as.Date(weird_df$DischargeDate,origin ="1970-01-01")  

#patient_loc_ID_long1 <- patient_loc_ID_long1[-c(58,201,236,288,250),] # remove none-sense dates
patient_loc_ID_long1$HCF <- trimws(patient_loc_ID_long1$HCF)
patient_loc_ID_long1 <- rbind(patient_loc_ID_long1, weird_df) %>%
  filter(AdmitDate>as.Date("2019-01-01") & DischargeDate>as.Date("2019-01-01")) %>%
  group_by(NCBI_SAMN, Patient_ID,variable) %>%
  arrange(NCBI_SAMN, variable) 

# collapse patient transfers that were from the same hospital
patient_loc_ID_long1$duplicate_HCF <- 0
for(i in 1:(nrow(patient_loc_ID_long1)-1)){
  if(patient_loc_ID_long1$NCBI_SAMN[i]==patient_loc_ID_long1$NCBI_SAMN[i+1] & patient_loc_ID_long1$HCF[i]==patient_loc_ID_long1$HCF[i+1]){ # if the previous transfer were the same as the latter transfer
   patient_loc_ID_long1$AdmitDate[i] <- patient_loc_ID_long1$AdmitDate[i+1] # Assign the Admit date as the previous transfer
    patient_loc_ID_long1$duplicate_HCF[i+1] <- 1
  }
  else{
    next
  }
}

patient_loc_ID_long1 <- patient_loc_ID_long1 %>%
  filter(duplicate_HCF==0)
patient_loc_ID_long1$HCF <- trimws(patient_loc_ID_long1$HCF)
```

```{r}
patient_loc_ID_wide <- patient_loc_ID_long1 %>%
  pivot_wider(id_cols = c(Patient_ID, NCBI_SAMN), values_from = HCF, names_from = variable, names_prefix = "HCF_exposure")

first_pt_loc <- patient_loc_ID_wide %>% 
  merge(patient_loc_key,by = c("Patient_ID","NCBI_SAMN"), all.x = TRUE) %>%
  select(-c(HCF_exposure_1,HCF_exposure_2,HCF_exposure_3,HCF_exposure_4,HCF_exposure_5,HCF_exposure_6,HCF_exposure_7)) %>%# delete redundent columns
  mutate(Specimen_collection_date = HCF_1_discharge_dt) %>%
  group_by(Patient_ID) %>%
  arrange(Patient_ID) %>%
  mutate(num = row_number()) %>%
  filter(num ==1) %>% # only keep first isolate for patients 
  ungroup()%>%
  arrange(Specimen_collection_date) # sort data by Specimen collection date
```

```{r keep first isolate}
# only keep first isolate for patients
# first_pt_loc <- patient_loc_key_ID %>%
#   group_by(Patient_ID) %>%
#   arrange(Patient_ID,Specimen_collection_date)%>%
#   mutate(num = row_number()) %>%
#   filter(num ==1) %>% # only keep first isolate for patients 
#   ungroup()%>%
#   arrange(Specimen_collection_date) # sort data by Specimen collection date

first_pt_loc <- first_pt_loc %>%
  filter(NCBI_SAMN!="SAMN16828101") # excluding the outgroup isolate

first_pt_loc_long <- patient_loc_ID_long1 %>%
  filter(NCBI_SAMN%in%first_pt_loc$NCBI_SAMN) # filter long form patient trace to the first isolated sample per patient


```


```{r}
get_epi_links <-function(first_pt_loc_long, seq_mat){ # input: 1. patient_transfer data frame with HCF exposures, unique ID and transfer dates. 2. shared_variant_matrix with sample unique ID
  
  # unit test
  #sample_list_long[[1]]
  #seq_mat = dna_shared_mat
  
  first_iso_df <- first_pt_loc_long %>%
    filter(!is.na(CollectionDate)) %>% # only keep obs. with sample collection dates 
    select(NCBI_SAMN, Patient_ID, CollectionDate, variable) %>%
    arrange(CollectionDate)
  #  keep only shared variants from the first isolate of each patient
  first_seq_mat <- seq_mat[which(row.names(seq_mat)%in%first_iso_df$NCBI_SAMN),which(colnames(seq_mat)%in%first_iso_df$NCBI_SAMN)] 
  first_seq_mat1 <- reorder_mat(first_seq_mat,first_iso_df$NCBI_SAMN) # reorder the matrix by the order of the NCBI_SAMN (unique ID) from first_iso_df
  
  # transpose first_seq_mat1 to long form 
  first_seq_mat1_long <- t(combn(colnames(first_seq_mat1),2))
  first_seq_mat1_long <- data.frame(first_seq_mat1_long, var = first_seq_mat1[first_seq_mat1_long])
  
  ## getting the maximal shared variants at each given time point
  dist_df <- NULL
  for (i in (2:nrow(first_seq_mat1))){ # start from second isolate (bc first isolate is automatically source)
    subset_v <- colnames(first_seq_mat1)[1:i]
    submat_long <- first_seq_mat1_long[which(first_seq_mat1_long$X1%in%subset_v&first_seq_mat1_long$X2%in%subset_v), ]
    submat_long$id <- i
    dist_df <- rbind(dist_df, submat_long)
  }
  
  # clean data.frame to only keep unique variants values per pair and rank'em
  dist_df <- dist_df %>%
    mutate(source = X1,
           NCBI_SAMN = X2) %>%
    distinct(source, NCBI_SAMN, var, .keep_all = TRUE) %>% # de-duplicate data
    #distinct() %>%
    group_by(id) %>%
    mutate(msv_rank = dense_rank(-var)) # get the rank of the variants by descending order
  
 
    dist_df_max <- dist_df %>%
      group_by(id) %>%
      filter(msv_rank == 1)
    
    # full join test_df with the HCF info data frame
    pt_transfer_long <- merge(dist_df_max,first_pt_loc_long, by.x = "X2",  by.y = "NCBI_SAMN",all = TRUE) # X2 would be the acquirer and X1 would be the source
    pt_transfer_long$CollectionDate <- as.numeric(pt_transfer_long$CollectionDate)
    pt_transfer_long$AdmitDate <- as.numeric(pt_transfer_long$AdmitDate)
    pt_transfer_long$CollectionDate <- as.numeric(pt_transfer_long$CollectionDate)
    pt_transfer_long$NCBI_SAMN <- pt_transfer_long$X2 # rename X2 to NCBI_SAMN
    #test_patient_transfer_long <- as.data.frame(test_patient_transfer_long)
    
    # subset to unique sources for each patient
    # pt_transfer_long1 <- pt_transfer_long1 %>%
    #   #arrange(NCBI_SAMN, CollectionDate) %>%
    #   distinct(source, NCBI_SAMN, HCF, .keep_all = TRUE) # keep only unique sources for each patient
    
    pt_transfer_long1 <- pt_transfer_long %>%
      group_by(Patient_ID, NCBI_SAMN) %>%
      arrange(Patient_ID,NCBI_SAMN, variable) %>%
      mutate(HCF_acquired = ifelse(CollectionDate>AdmitDate + 2, 1, 0))  # find isolates where collection date is <48 hrs after admit date
    
    # get cases that only have one healthcare exp. and was collected less than 48 hours after admission
    single_HCF_and_non_acquirer <- pt_transfer_long1 %>%
      distinct(Patient_ID, NCBI_SAMN, HCF, .keep_all = TRUE) %>%
      group_by(Patient_ID) %>%
      summarise(num_of_HCFs=n())
    
    # flag isolates that have same-day collection HCF, and only single HCF exp. 
    pt_transfer_long1 <- pt_transfer_long1 %>%
      mutate(HCF_sameday_and_single = ifelse(HCF_acquired==0 & Patient_ID %in%single_HCF_and_non_acquirer$Patient_ID[which(single_HCF_and_non_acquirer$num_of_HCFs==1)],1,0))
    
    ## Populate var specifying putative NDM positive NDM
    # If Isolate was collectied>48hr, then positive HCF would be sample collection HCF, if <48hrs, then positive HCF would be the HCF prior to sample collection
    # pt_transfer_long1$positive_HCF <- 0
    #  for(i in 1:nrow(pt_transfer_long1)){
    #    pt_transfer_long1$CollectionDate[i]
    #    #pt_transfer_long1$Discharge[i]
    #    id_v <- pt_transfer_long1$NCBI_SAMN[i]
    # 
    #    if(is.na(pt_transfer_long1$HCF_acquired[i])){
    #      next
    #    }
    #    else if (pt_transfer_long1$HCF_acquired[i]==0 & pt_transfer_long1$HCF_sameday_and_single[i]==0){
    #      pt_transfer_long1$positive_HCF[i] <- unique(pt_transfer_long1$HCF[pt_transfer_long1$NCBI_SAMN==id_v & pt_transfer_long1$variable==2])
    #    }
    #    else{
    #      pt_transfer_long1$positive_HCF[i] <- unique(pt_transfer_long1$HCF[pt_transfer_long1$NCBI_SAMN==id_v & pt_transfer_long1$variable==1])
    #    }
    #  }

    # # 4. Assigning transfer groups ######
    pt_transfer_long1$epi_type <- 0
    pt_transfer_long1$epi_HCF <- 0
    pt_transfer_long1$from <- 0
    pt_transfer_long1$to <- 0
 
    for(i in 1:nrow(pt_transfer_long1)){
      
      # acquirer patient of interest info
      v_aq <- pt_transfer_long1$NCBI_SAMN[i] # the name of the acquirer to compare
      HCF_aq <- pt_transfer_long1$HCF[i] # the HCF to compare for shared exposure 
      #posititive_HCF_aq <- pt_transfer_long1$HCF_acquired[i] # whether this this isolate is sampled at HCF of acquirement (1/0/NA - for non- sample collection HCFs) 
      
      # source patient info
      v_source <- pt_transfer_long1$source[i] # name of the source
      ls_HCF_source <- pt_transfer_long1$HCF[which(pt_transfer_long1$NCBI_SAMN==v_source)] #sources' healthcare exps.
      dt_source <- unique(pt_transfer_long1$HCF[which(pt_transfer_long1$NCBI_SAMN==v_source &!is.na(pt_transfer_long1$CollectionDate))]) # source' sample collection HCF
      #dt_source <- unique(pt_transfer_long1$HCF[which(pt_transfer_long1$NCBI_SAMN==v_source &!is.na(pt_transfer_long1$CollectionDate))]) # source' sample collection HCF
      dt_aq <- unique(pt_transfer_long1$HCF[which(pt_transfer_long1$NCBI_SAMN==v_aq & !is.na(pt_transfer_long1$CollectionDate))]) # acquirers' sample collection HCF
      
      # skip index patient who does not have a source
      if(is.na(v_source)){
        next
      }
      
      ## If shared exp. isolate collection were <48hrs before admission
      else if(pt_transfer_long1$HCF_acquired[i]==0 & pt_transfer_long1$HCF_sameday_and_single[i]!=1 & !is.na(pt_transfer_long1$HCF_acquired[i])){
        next
      }
      
      ## If there WERE shared exp. between source and acquirer
       else if(HCF_aq %in% ls_HCF_source){
          if(HCF_aq==dt_source & !is.na(pt_transfer_long1$CollectionDate[i])){ # if shared HCF is the samples collection HCF of source & is the samples collection HCF of the patient of interest
            pt_transfer_long1$epi_type[i] <- 1
            pt_transfer_long1$epi_HCF[i] <- dt_source
            pt_transfer_long1$from[i] <- dt_source
            pt_transfer_long1$to[i] <- HCF_aq
            # dt_source==HCF_aq
          }
          else if(!is.na(pt_transfer_long1$CollectionDate[i])){ # if shared HCF is the collection HCF of the patient of interest
            pt_transfer_long1$epi_type[i] <- 2
            pt_transfer_long1$epi_HCF[i] <-  dt_source
            pt_transfer_long1$from[i] <- HCF_aq
            pt_transfer_long1$to[i] <-  HCF_aq
            #from: shared health exp. with source (HCF_aq) to:  source sample collection HCF (dt_source)
          }
          else if(HCF_aq==dt_source){ # if the shared HCF is the sample collection HCF of source
            pt_transfer_long1$epi_type[i] <- 3
            pt_transfer_long1$epi_HCF[i] <-  HCF_aq
            pt_transfer_long1$from[i] <- dt_source
            pt_transfer_long1$to[i] <- dt_aq
          }
          else { # if the shared HCF is neither sample collection HCF of source nor patient of interest 
            pt_transfer_long1$epi_type[i] <- 4
            pt_transfer_long1$epi_HCF[i] <-  HCF_aq
            pt_transfer_long1$from[i] <- HCF_aq
            pt_transfer_long1$to[i] <- dt_aq
          }
        }
      ## if there were NO shared exp. between source and acquirer
        else {
          pt_transfer_long1$epi_type[i] <- 0
          pt_transfer_long1$epi_HCF[i] <- dt_source
          pt_transfer_long1$from[i] <- dt_source
          pt_transfer_long1$to[i] <- dt_aq
        }
      }
    
    pt_transfer_long1$epi_type <- as.numeric(pt_transfer_long1$epi_type)
    pt_transfer_long1 <- pt_transfer_long1 %>%
      filter(from!=0)
    
    ## if isolates were has the same from. and to. HCF, and are epi_type==2/3 then reassign epi_type to 1
     # pt_transfer_long1 <- pt_transfer_long1 %>%
     #   mutate(epi_type_new = ifelse(from==to & (epi_type==3|epi_type==4),1,epi_type),
     #          epi_type=epi_type_new) %>%
     #   select(-epi_type_new)
    
    pt_transfer_long1 <- pt_transfer_long1 %>%
      filter(!is.na(source)) # remove index case
  
    
    # updating the source' source status (source') to reflect new source if source used to have unknown source assignments
    ## NCBI_SAMN ids that are assigned to respective epi_types
  
    
    epi_2_0_df <- pt_transfer_long1 %>%
      distinct(NCBI_SAMN, epi_type) %>%
      group_by(NCBI_SAMN) %>%
      mutate(n = n(),
             epi_2_0 = ifelse(n()==2 & sum(epi_type)==2,1,0))
    
## summarise the epitype to find unknown epi_types
    pt_transfer_long1_sum <- pt_transfer_long1 %>%
      mutate(epi_type = ifelse(is.na(epi_type),0,epi_type)) %>%
      group_by(NCBI_SAMN) %>%
      arrange(NCBI_SAMN,epi_type) %>%
      summarise(sum = sum(epi_type)) # if sum==0 means this isolate only have unknown assignments
    
    other_list <- pt_transfer_long1_sum$NCBI_SAMN[which(pt_transfer_long1_sum$sum==0)] ## genomic only links
    #intra_list <- unique(pt_transfer_long1$NCBI_SAMN[which(pt_transfer_long1$epi_type==1)]) ## intra-facility transmission
    
    
      pt_transfer_long3 <- pt_transfer_long1 %>%
        group_by(NCBI_SAMN) %>%
        arrange(CollectionDate,NCBI_SAMN) %>%
        mutate(row_num = row_number()) %>%
        mutate(epi_type = ifelse(NCBI_SAMN%in%other_list & !is.na(CollectionDate),5,ifelse(NCBI_SAMN%in%other_list & row_num==1,5,epi_type))) %>%  # flag the isolate that only has inter-unknown
        filter(epi_type!=0) %>% # clean out uninformative data
        group_by(NCBI_SAMN) %>%
        arrange(NCBI_SAMN,epi_type) %>%
        distinct(NCBI_SAMN, source, X1, X2, epi_type, epi_HCF, from , to, .keep_all = TRUE)
    
    epi_2_source_unknown <- other_list[other_list%in%unique(pt_transfer_long3$source[which(pt_transfer_long3$epi_type==2)])] 
    epi2_df <- data.frame(NCBI_SAMN = pt_transfer_long3$NCBI_SAMN[pt_transfer_long3$epi_type==2],
                          source = pt_transfer_long3$source[pt_transfer_long3$epi_type==2]) # find the sources isolate id (NCBI_SAMN) for isolated that are epi_type 2
    source_to_update <- unique(pt_transfer_long3$NCBI_SAMN[pt_transfer_long3$NCBI_SAMN%in%epi2_df$source & pt_transfer_long3$epi_type==5]) # find the ones that only have unknown sources
    #intersect(source_to_update,other_list)
    
    epi2_df_sub <- epi2_df %>%
      filter(source%in%source_to_update)
    source_update_df <- pt_transfer_long3[pt_transfer_long3$source%in%epi2_df_sub$source & pt_transfer_long3$NCBI_SAMN %in% epi2_df_sub$NCBI_SAMN & pt_transfer_long3$epi_type==2,] # subset to epi_type2 sources that has unknown sources 
 
    pt_transfer_long3$update <- 0 # add variable to indicate original
    source_update_df$update <- 1 # add variable to indicate updated
    source_update_df <- source_update_df %>%
      ungroup() %>%
      mutate(new_NCBI_SAMN = source,
             source = NCBI_SAMN,
             from = HCF,
             to = epi_HCF) %>%
      select(-NCBI_SAMN) %>%
      mutate(NCBI_SAMN = new_NCBI_SAMN) %>%
      select(-new_NCBI_SAMN) 
    
    pt_transfer_long4 <- rbind(pt_transfer_long3, source_update_df) # bind with updated df
    table(pt_transfer_long4$epi_type)
  
  
  out <- pt_transfer_long4 # output list:1. pair_df for getting sources with maximum shared variants 2. and the actual output of interest
  #pair_df var: V1(destination isolate), V2(max_shared_variants, aka potential source)
  
  return(out)
  
}
```

```{r}
# test if function works
#dna_shared_mat1 <- dna_shared_mat[rownames(dna_shared_mat)!="SAMN16828101",colnames(dna_shared_mat)!="SAMN16828101"]
output <- get_epi_links(first_pt_loc_long = first_pt_loc_long , seq_mat = dna_shared_mat)
#output_test2 <- get_epi_links(first_pt_loc = first_pt_loc , seq_mat = dna_shared_mat)
#out <- get_epi_links(first_pt_loc = first_pt_loc , seq_mat = dna_shared_mat)
MDHHS_test <- output

## test epi_type2 
## get the sample IDs that only have epitype2 linkage or epi_type==0 linkage
#pt_transfer_long1 <- output[[4]]
# pair_df <- output[[2]]
# #MDHHS_max_shared_var <- output[[2]]
# MDHHS_first_mat <- output[[1]]

#MDHHS_max_shared_var1 <- MDHHS_max_shared_var[-1] # to exclude the first element from list (bc this guy don't have a source)
#detach(package:seqinr)
## unlist MDHHS_max_shared_val to a df
#max_shared_val_df <- Reduce(full_join,MDHHS_max_shared_val) # combine all dataframes from the list in to a datafrmae

MDHHS_epitype <- MDHHS_test %>%
  select(NCBI_SAMN, source, epi_type,epi_HCF) # subset data to the just the merge identifiers and epitype 

#max_shared_val_df1 <- merge(max_shared_val_df, MDHHS_epitype,by.x=c("id","source"), by.y=c("NCBI_SAMN","source")) ## merge epitype information of both id and source together
# max_shared_val_df1 <- merge(pair_df,MDHHS_epitype,by.x=c("X2","X1"), by.y=c("NCBI_SAMN","source")) %>%
#   filter(msv_rank==1)

#colnames(MDHHS_test)
MDHHS_test$HCF <- trimws(MDHHS_test$HCF)
MDHHS_sum <- MDHHS_test %>%
   group_by(NCBI_SAMN, source) %>%
   summarise(n = n()) %>%
   group_by(NCBI_SAMN) %>%
   mutate(row_num = row_number(),
          count = max(row_num))

```

```{r unknown proportion}
 # get proportion of unknown to known
MDHHS1 <- MDHHS_test %>%
  group_by(Patient_ID) %>%
  mutate(n = n(), # how many transfer types each patient have
         weight = 1/n, # weights so that all transfer types from one patient would sum up to 1
         known = ifelse(epi_type == 5, 0,1))%>% # we are only interested in whether the transfer types are known or unknown
  ungroup() %>%
  group_by(known)%>% # group by the known
  summarize(count = sum(weight)) # get sum of patients with unknown info and known info respectively

# calculate the proportion of unknown/(unknown+known)

MDHHS_unknown_prop <-  MDHHS1$count[MDHHS1$known==0]/(MDHHS1$count[MDHHS1$known==0]+MDHHS1$count[MDHHS1$known==1])
```

```{r permutation function}
########### function: epi_perm_test ##################
epi_perm_test <- function(epi_df, n_perm){
  # Unit test variables
  #n_perm = 10
  #epi_df <- first_iso_df
  startTime <- Sys.time()
   # list to store each randomized data frame
  sample_list <- vector(mode = "list", length (n_perm)) 
  
  for(i in 1:n_perm){
    
      sample_list[[i]] <- epi_df # duplicate original first_pt_loc data 
      # randomize IDs
    sample_list[[i]]$NCBI_SAMN <- sample(sample_list[[i]]$NCBI_SAMN, size = nrow(sample_list[[i]]), replace = FALSE) # sample the NCBI_SAMN IDs of the data frame
      
      # match Isolate_ID to corresponding Patient_ID (Isolate ID should not be randomized, only randomizing the HCF locations)
    sample_list[[i]] <- sample_list[[i]] %>%
      ungroup() %>%
        select(-c(Patient_ID,Isolate_ID)) # delete the randomized Patient ID
    sample_list[[i]] <- merge(sample_list[[i]] , key3, by = "NCBI_SAMN") # merge key with NCBI_SAMN and Patient ID to get correct corresponding Patient IDs
  }
  
  
  sample_list_long <- lapply(sample_list, function(df){
    df <- setDT(df)
    df_long <- melt(df, id.vars = c("NCBI_SAMN"),
                                measure.vars = list(HCF = c(paste0("HCF_exposure",seq(1,7,1))),
                                                    AdmitDate = c(paste0("HCF_",seq(1,7,1),"_admit_dt")),
                                                    DischargeDate = c(paste0("HCF_",seq(1,7,1),"_discharge_dt")),
                                                    CollectionDate = "Specimen_collection_date",
                                                    Patient_ID = rep(c("Patient_ID"),7),
                                                    Isolate_ID = rep(c("Isolate_ID"),7)))
    
    out <- df_long %>%
      filter(HCF!="")
    out$HCF <- trimws(out$HCF)
    out$AdmitDate <- as.Date(out$AdmitDate, format = "%m/%d/%y")
    out$DischargeDate <- as.Date(out$DischargeDate, format = "%m/%d/%y")
    out$CollectionDate <- as.Date(out$CollectionDate, format = "%m/%d/%y")
    return(out)
  }
  )
  
  v_prop <- sapply(sample_list_long, function(df){
    # apply the get_epi_links function to permutation datasets
    epi_link_df <- get_epi_links(first_pt_loc = df , seq_mat = dna_shared_mat)
    # get the sum of patients with known/ unknown transfer epi_links respectively 
    epi_link_df1 <- epi_link_df %>%
      group_by(Patient_ID) %>%
      mutate(n = n(),
             weight = 1/n,
             known = ifelse(epi_type == 5, 0,1))%>%
      ungroup() %>%
      group_by(known)%>%
      summarize(count = sum(weight))
    
    # calculate proportion of unknown
    prop_unknown <- epi_link_df1$count[epi_link_df1$known==0]/(epi_link_df1$count[epi_link_df1$known==0]+ epi_link_df1$count[epi_link_df1$known==1])
    #print(prop_unknown)
    return(prop_unknown)
  })
  endTime <- Sys.time()
  print("Total run time")
  print(endTime - startTime)
  return(v_prop)
}

```

```{r run permutation function}
# set.seed(0)
# #sample_prop_unknown <- epi_perm_test(first_pt_loc, 1000)
# prop_unknown_df <- as.data.frame(sample_prop_unknown)
# data_prop <- length(which(MDHHS_test$epi_type==5))/nrow(MDHHS_test)
# ggplot(prop_unknown_df, aes(x = sample_prop_unknown))+ geom_histogram(binwidth = 0.02) + theme_bw() + labs(x = "Proportion no epidemiologic link", y= "Frequency") + geom_vline(xintercept = data_prop, color = "red",linetype = "dashed")
#ggsave("~/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/exploratory_figures/2024_09_09_permutation_test1000.pdf", width = 9)
# # testing normality of data
# qqnorm(sample_prop_unknown)
# qqline(sample_prop_unknown)
# shapiro.test(sample_prop_unknown)
# ks.test(sample_prop_unknown, 'pnorm')
# wilcox.test(sample_prop_unknown, mu = MDHHS_unknown_prop)
# z.test(sample_prop_unknown,
#   y = NULL,
#   alternative = "two.sided",
#   mu = MDHHS_unknown_prop,
#   sigma.x = sd(sample_prop_unknown),
#   sigma.y = NULL,
#   conf.level = 0.95
# )

```


```{r}
## empty data.frame
pt_trans_df <- data.frame(
  epi_type = MDHHS_test$epi_type,
  HCF1 = MDHHS_test$from,
  HCF2 = MDHHS_test$to,
  pt1 = MDHHS_test$source,
  pt2 = MDHHS_test$NCBI_SAMN,
  pt_id = MDHHS_test$X2,
  update = MDHHS_test$update,
  HCF_acquired = MDHHS_test$HCF_acquired
)


pt_trans_df$inter_date <- 0 # this is gonna be the variable for the epi_type==4 groups' Discharge date of when "acquirer" left the shared HCF & for epi_type==2 & update==0 groups' start date of when "acquirer" was hospitalized, and for epi_type==2 & update==1 "source" hospitalization start date 
pt_trans_df$pt1_date <- 0 ## The times when these transmission event occurs (start of the arrow)
pt_trans_df$pt2_date <- 0 ## The times when these transmission event occurs (end of the arrow)

## populate 
for(i in 1:nrow(pt_trans_df)){
  ## indexing the sample collection and of source and acquirers
  acq <- pt_trans_df$pt2[i] # acquirer index
  source <- pt_trans_df$pt1[i] # source index
  pt1_Collection_date_i <- which(patient_loc_ID_long1$NCBI_SAMN==source & !is.na(patient_loc_ID_long1$CollectionDate)) # index for the sample collection HCF , collection date
  #print(c(i, pt1_Collection_date_i))
  pt2_Collection_date_i <- which(patient_loc_ID_long1$NCBI_SAMN==acq & !is.na(patient_loc_ID_long1$CollectionDate)) # index for the sample collection HCF, collection date
  #print(c(i, pt2_Collection_date_i))
  # 
  # if(pt_trans_df$epi_type[i]==2){
  #  pt_trans_df$pt1_date[i] <- patient_loc_ID_long1$AdmitDate[which(patient_loc_ID_long1$NCBI_SAMN==source & patient_loc_ID_long1$HCF==pt_trans_df$HCF1[i])][1] # from the admission date of the shared HCF 
  #   pt_trans_df$pt2_date[i] <- patient_loc_ID_long1$CollectionDate[pt2_Collection_date_i] # to the sample collection date of the acquirer
  #   pt_trans_df$inter_date[i] <- patient_loc_ID_long1$AdmitDate[which(patient_loc_ID_long1$NCBI_SAMN==source & patient_loc_ID_long1$HCF==pt_trans_df$HCF1[i])][1] # the intermediate HCF date for for indirect shared HCF, if there is multiple share HCF collection dates then choose the latest one
  # } 

  if(pt_trans_df$epi_type[i]%in%c(2,4)){
    pt_trans_df$pt1_date[i] <- patient_loc_ID_long1$AdmitDate[which(patient_loc_ID_long1$NCBI_SAMN==source & patient_loc_ID_long1$HCF==pt_trans_df$HCF1[i])][1] # from the admission date of the shared HCF 
    pt_trans_df$pt2_date[i] <- patient_loc_ID_long1$CollectionDate[pt2_Collection_date_i] # to the sample collection date of the acquirer
    pt_trans_df$inter_date[i] <- patient_loc_ID_long1$AdmitDate[which(patient_loc_ID_long1$NCBI_SAMN==source & patient_loc_ID_long1$HCF==pt_trans_df$HCF1[i])][1] # the intermediate HCF date for for indirect shared HCF, if there is multiple share HCF collection dates then choose the latest one
  }
  else {
    pt_trans_df$pt1_date[i] <- patient_loc_ID_long1$CollectionDate[pt1_Collection_date_i]
    pt_trans_df$pt2_date[i] <- patient_loc_ID_long1$CollectionDate[pt2_Collection_date_i]
  }
}

```

```{r}
test_pt_node <- data.frame(pt = c(pt_trans_df$pt1,pt_trans_df$pt2),
                           CollectionDate = c(pt_trans_df$pt1_date, pt_trans_df$pt2_date),
                           HCF = c(pt_trans_df$HCF1, pt_trans_df$HCF2)) 

# subset patient transfer data to the collection date and HCF and corresponsing isolates
iso_collection_date_HCF <- patient_loc_ID_long1 %>%
  filter(!is.na(CollectionDate)) %>%
  select(NCBI_SAMN, CollectionDate) %>%
  mutate(pt = NCBI_SAMN,
         CollectionDate = as.numeric(CollectionDate))
## Using intersect to find the isolates that have matching pt and collection dates in the 2 datasets 
collectiondate_df <- intersect(iso_collection_date_HCF[,c("pt","CollectionDate")],test_pt_node[,c("pt","CollectionDate")])
collectiondate_df$type <- "patient"

# full join with the collectiondate_df, this way the ones that are not sample collection dates will be NA in the type column
test_pt_node <- merge(collectiondate_df, test_pt_node,  by.x = c("pt", "CollectionDate"),by.y = c("pt","CollectionDate") , all = TRUE)
# recode NA values in the "type" column the to reflect the HCF
test_pt_node <- test_pt_node %>%
  mutate(type = ifelse(is.na(type),"HCF",type))

test_pt_node <- test_pt_node %>%
  arrange(CollectionDate)%>%
  mutate(pt = ifelse(type=="HCF", HCF,pt )) %>%
  group_by(pt) %>%
  mutate(pt = ifelse(type=="HCF", paste0(pt, "_",row_number()),pt))%>%
  group_by(HCF) %>%
  mutate(id = cur_group_id(),
         y = id,
         x = as.numeric(CollectionDate))

pt_id_df <- test_pt_node 
  
```


```{r}
### pt_id_df: the patient nodes #####
# pt_id_df <- data.frame(
#   pt = unique(c(pt_trans_df$pt1,pt_trans_df$pt2)), # getting all the patients isolates in the network as nodes
#   type = "patient") # assign as patient nodes
# 
# for ( i in 1:nrow(pt_id_df)){
#   id <- which(patient_loc_ID_long1$NCBI_SAMN== pt_id_df$pt[i] & !is.na(patient_loc_ID_long1$CollectionDate)) # find the row that contains CollectionDate information
#   if(is.na(pt_id_df$pt[i])){
#     next()
#   }
#   pt_id_df$HCF[i] <- patient_loc_ID_long1$HCF[id] # assign that healthcare facility to (sample collection HCF) the patient
#   pt_id_df$CollectionDate[i] <- patient_loc_ID_long1$CollectionDate[id] # assgined that Collection time to the patient
# }


```

```{r}
# ## add the HCFs from epitype_4 for the nodes in the graph
# HCF_epi <- data.frame(pt = pt_trans_df$HCF1[pt_trans_df$inter_date!=0], # assign the source HCF1 that are epi_type==4 (in other words the ones with a inter_date)
#                        HCF= pt_trans_df$HCF1[pt_trans_df$inter_date!=0], # assign the source HCF1 that are epi_type==4 (in other words the ones with a inter_date)
#                        CollectionDate =pt_trans_df$inter_date[pt_trans_df$inter_date!=0], # assign CollectionDate as the epi_type==4 date (when the case was discharged from this particular HCF)
#                        type = "HCF") # assign type "HCF" to distinguish from "patient"
# 
# HCF_epi <- HCF_epi %>%
#   distinct()
# pt_id_df <- rbind(pt_id_df, HCF_epi) # bind these together
# 
# # assign unique id to each HCF
# pt_id_df <- pt_id_df %>%
#   group_by(HCF) %>%
#   mutate(id=cur_group_id()) # assign unique id to each HCF
# 
# 
# ## color palette
# HCF_col <- c("#ed9ec3", "#FF7F00", "#FFD400", "#FFFF00", "#BFFF00", "#17e364", "#00EAFF", "#0095FF", "#0040FF", "#AA00FF", "#ff4c7c", "#EDB9B9", "#E7E9B9", "#B9EDE0", "#B9D7ED", "#DCB9ED", "#8F2323", "#8F6A23", "#4F8F23", "#23628F", "#6B238F", "#fceaae","#FFA405","#C20088","#E0FF66" )
# 
# pt_id_df <- pt_id_df %>%
#   group_by(HCF) %>%
#   mutate(y = cur_group_id(), # y axis values (unique identifiers of  HCFs)
#          x = CollectionDate) %>% # x axis values
#   ungroup() %>%
#   arrange(CollectionDate) %>%
#   group_by(pt) %>%
#   #distinct(pt,type, HCF, CollectionDate, .keep_all = TRUE) %>%
#   mutate(pt = ifelse(type=="HCF",paste0(pt,"_",row_number()),pt)) ## add a unique identifier for each source HCF for epi_type 4 (so that they don't all link back to the same one)

```

```{r}
pt_trans_df_edge <- pt_trans_df %>%
  mutate(old_pt1 = pt1) %>%
  mutate(pt1 = ifelse(inter_date!=0, HCF1,pt1)) %>%# substitute pt1 to HCF for epi_type4  (source HCF)
  ungroup() %>%
  group_by(pt1) %>% 
  arrange(pt1_date) %>%
  mutate(pt1 = ifelse(inter_date!=0,paste0(pt1,"_",row_number()),pt1 )) %>% # create unique identifiers for each transfer event per HCF for epi_type4 (We do this because otherwise the networks would all be linked to a same HCF because they all have the same names -aka no unique identifiers per transmission event)
  mutate(from = pt1, 
         to = pt2) %>%
  distinct()

```

```{r, fig.show='hide'}
trans_edge_test3 <-  pt_trans_df_edge %>%
  select(epi_type, HCF1, HCF2, pt1, pt2, pt1_date, pt2_date, from, to, old_pt1) %>%
  group_by(pt2, HCF1, HCF2) %>%
  mutate(pt_link_n= n(), ## count the numbers of HCF links in regard to one patient
         n_id = cur_group_id(),## assign group ids to these groups
         genomic_only = ifelse(epi_type==5,1,0)) %>% 
  group_by(pt2) %>%
  mutate(n=n(), ## count the numbers of time a patient was a assigned to a group
         freq = pt_link_n/n, ## count the percentage of the times groups were being assigned per patient
         ## create arbitrary confidece level variable
         CL = case_when(epi_type!=5 & freq >0.5 ~"Genomic & Epi",
                        epi_type==5 & freq >0.5 ~"Genomic only",
                        TRUE ~ "Uncertain"))  

### plot MSV of the predicted transmission events colored by confidence
pt_trans_df_edge_msv <- trans_edge_test3
pt_trans_df_edge_msv$msv <- 0 
pt_trans_df_edge_msv$snv <- 0 

for (i in unique(pt_trans_df_edge_msv$old_pt1)){
  for(j in unique(pt_trans_df_edge_msv$pt2)){
    pt_trans_df_edge_msv$msv[pt_trans_df_edge_msv$old_pt1==i & pt_trans_df_edge_msv$pt2==j] <- dna_shared_mat[i,j]
  }
}
# 
# for (i in unique(pt_trans_df_edge_msv$old_pt1)){
#   for(j in unique(pt_trans_df_edge_msv$pt2)){
#     pt_trans_df_edge_msv$snv[pt_trans_df_edge_msv$old_pt1==i & pt_trans_df_edge_msv$pt2==j] <- dist_mat[i,j]
#   }
# }

pt_trans_df_edge_msv <- pt_trans_df_edge_msv %>%
  mutate(CL = ifelse(HCF1==HCF2 & CL!="Uncertain", "intra",CL))

ggplot(pt_trans_df_edge_msv) + geom_histogram(aes(x = snv)) +facet_grid(rows = vars(CL))

ggplot(pt_trans_df_edge_msv)+ geom_point(aes(x = snv, y =HCF1),color = "#F4C434") + geom_point2(aes(x = snv, y = HCF2),color = "#9ACC8F") + geom_segment(aes(y=HCF1,yend = HCF2, x = snv, xend = snv, color = CL), alpha =.3)

sum_HCF_linkage <- trans_edge_test3 %>%  # percentage of times patients were assigned to only one HCF
  group_by(n_id) %>%
  mutate(rows = row_number()) %>%
  filter(rows==1)

kruskal.test(snv~CL, data = pt_trans_df_edge_msv)

################done ##########

ggplot(sum_HCF_linkage, aes(x= freq*100, fill = as.factor(epi_type))) + geom_bar() +labs(fill ="transmission type") +scale_fill_manual( values = c("1"="#F4C434","2"="#9ACC8F","3"="#AF4474","4"="skyblue","5"="#2D4B65"), labels = c("1"="intra-facility","2"="inter-facility (acquirer)", "3"="inter-facility (source)", "4"="inter-facility (shared)", "5"="inter-facility (unknown)")) + theme_bw() + theme(panel.grid = element_blank()) + geom_vline(xintercept = 50,linetype =2, color ="red" )
#ggsave("~/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/CL_dist_hist.pdf", width = 7, height = 5)
```

```{r, fig.show='hide'}
########reformat the patient network edges####################
pt_trans_df_edge <- pt_trans_df %>%
  mutate(old_pt1 = pt1) %>%
  mutate(pt1 = ifelse(inter_date!=0, HCF1,pt1)) %>%# substitute pt1 to HCF for epi_type4  (source HCF)
  ungroup() %>%
  group_by(pt1) %>% 
  arrange(pt1_date) %>%
  mutate(pt1 = ifelse(inter_date!=0,paste0(pt1,"_",row_number()),pt1 )) %>% # create unique identifiers for each transfer event per HCF for epi_type4 (We do this because otherwise the networks would all be linked to a same HCF because they all have the same names -aka no unique identifiers per transmission event)
  mutate(from = pt1, 
         to = pt2) %>%
  distinct()


pt_id_df_test <- pt_id_df %>%
  filter(x%in% unique(c(pt_trans_df_edge$pt1_date,pt_trans_df_edge$pt2_date)))
######## reformatting patient network edge ends ################
pt_net_test <- tbl_graph(pt_id_df, pt_trans_df_edge , directed = TRUE) %>% activate(nodes) ## create graph
# 
# 
# pt_net_test %>% ggraph(x=x,y=y) + geom_node_point(size = 2, aes(color = HCF, shape=type)) + geom_edge_fan(aes(edge_colour = as.factor(epi_type)), arrow = arrow(length = unit(1.5, 'mm'), type = "closed"), end_cap = circle(1, 'mm'), alpha =0.4)  + scale_edge_color_manual(name = "transfer type", values = c("1"="#F4C434","2"="#9ACC8F","3"="#AF4474","4"="skyblue","5"="#2D4B65"), labels = c("1"="intra-facility","2"="inter-facility (acquirer)", "3"="inter-facility (source)", "4"="inter-facility (shared)", "5"="inter-facility (unknown)"))+ theme(legend.position = "none") + scale_color_manual(values = location_col) + scale_shape_manual(values = c(17, 19))

## Only keeping one transmission event where the source to that particular patient is the same
trans_edge_test2 <- pt_trans_df_edge_msv %>%
  group_by(epi_type, pt2, HCF1) %>% # finding the same healthcare facility where NDM transmitted to the same  pt2 
  arrange(pt1_date) %>% # sort by the dates of the source so the earliest one will come first
  mutate(shared_HCF = row_number(), 
         to_delete = ifelse( shared_HCF!=1,1,0)) %>% ## only keeping first observed tranmission event  
  filter(to_delete==0) #%>% 
#  filter(epi_type!=5) ## filter out epitype5 bc it makes it messy
pt_net_test2 <- tbl_graph(pt_id_df, trans_edge_test2 , directed = TRUE) %>% activate(nodes) # create graph

my_layout <- create_layout(pt_net_test, 'lgl')

#### HCF layout order ######
inter_trans <-pt_trans_df %>%
  filter(HCF1!=HCF2) # only keep interfaciltiy transfers 
# How many times a HCF with connected to another HCF

HCF_freq <- data.frame(HCF = c(pt_trans_df$HCF1,pt_trans_df$HCF2)) %>%
  group_by(HCF) %>%
  summarise(n=n())



HCF_freq1 <- HCF_freq[match(c("Ascension Macomb Oakland - Madison Heights Campus",
                              "Ascension Providence Park Southfield",
                              "Select Specialty Hospital Grosse Pointe",
                              "Karmanos Cancer Center",
                              "Henry Ford Hospital",
                              "Lake Huron Medical Center",
                              "Riverview Health & Rehab Center Jefferson",
                              "Beaumont Royal Oak",
                              "DMC Sinai Grace Hospital",
                              "Ascension St. John",
                              "Vibra DMC Campus",
                              "Select Specialty Hospital Northwest Detroit",
                              "Regency at Chene",
                              "Select Specialty Hospital Downriver",
                              "Michigan Medicine",
                              "McLaren Port Huron",
                              "DMC Harper University Hospital",
                              "DMC Detroit Receiving Hospital",
                              "Villa at Rose City",
                              "DMC Rehabilitation Institute of Michigan",
                              "Beaumont Taylor",
                              "John D. Dingell Department of Veterans Affairs Medical Center",
                              "Divinity Home Care",
                              "Ascension Providence Novi",
                              "Ascension Macomb Oakland - Warren Campus",
                              "Father Murray, A Villa Center",
                              "Omni Continuing Care"),HCF_freq$HCF),] %>%
  mutate(row_id = row_number())

# change the order of the y axis
pt_id_df1 <- pt_id_df 
for (i in 1:nrow(pt_id_df1)){
  if(!(pt_id_df1$HCF[i]%in%HCF_freq1$HCF)){
    next
  }
  pt_id_df1$y[i] <- HCF_freq1$row_id[which(HCF_freq1$HCF==pt_id_df1$HCF[i])] 
}

###############################

pt_net_test3 <- tbl_graph(pt_id_df1, trans_edge_test2 , directed = TRUE) 
#detach(package:seqinr)
pt_net_test3 %>% ggraph(x=x,y=y) + geom_edge_fan(aes(edge_colour = as.factor(epi_type)), arrow = arrow(length = unit(1.5, 'mm'), type = "closed"), end_cap = circle(1, 'mm'), alpha =0.4) + geom_node_point(size = 2, aes(color = HCF, shape=type)) + scale_edge_color_manual(name = "transfer type", values = c("1"="#F4C434","2"="#9ACC8F","3"="#AF4474","4"="skyblue","5"="#2D4B65"), labels = c("1"="intra-facility","2"="inter-facility (acquirer)", "3"="inter-facility (source)", "4"="inter-facility (shared)", "5"="inter-facility (unknown)"))+theme_graph(my_layout) +theme(legend.position = "none") +scale_color_manual(values = location_col)+ scale_shape_manual(values = c(17, 19))

node_ls <- pt_net_test3%>%activate(nodes) %>%as.data.frame()
edge_ls <- pt_net_test3%>%activate(edges)%>% as.data.frame()

```
> Filter for more likely single facility
Supplemental Figure 1a

```{r,fig.keep = "last"}
trans_edge_test3 <-  pt_trans_df_edge %>%
  select(epi_type, HCF1, HCF2, pt1, pt2, pt1_date, pt2_date, from, to, old_pt1) %>%
  group_by(pt2, HCF1, HCF2) %>%
  mutate(pt_link_n= n(), ## count the numbers of HCF links in regard to one patient
         n_id = cur_group_id(),## assign group ids to these groups
         genomic_only = ifelse(epi_type==5,1,0)) %>% 
  group_by(pt2) %>%
  mutate(n=n(), ## count the numbers of time a patient was a assigned to a group
         freq = pt_link_n/n, ## count the percentage of the times groups were being assigned per patient
         ## create arbitrary confidece level variable
         CL = case_when(epi_type!=5 & freq >0.5 ~"Genomic & Epi",
                        epi_type==5 & freq >0.5 ~"Genomic only",
                        TRUE ~ "Uncertain"))  

### plot MSV of the predicted transmission events colored by confidence
pt_trans_df_edge_msv <- trans_edge_test3
pt_trans_df_edge_msv$msv <- 0 
pt_trans_df_edge_msv$snv <- 0 

for (i in unique(pt_trans_df_edge_msv$old_pt1)){
  for(j in unique(pt_trans_df_edge_msv$pt2)){
    pt_trans_df_edge_msv$msv[pt_trans_df_edge_msv$old_pt1==i & pt_trans_df_edge_msv$pt2==j] <- dna_shared_mat[i,j]
  }
}
# 
# for (i in unique(pt_trans_df_edge_msv$old_pt1)){
#   for(j in unique(pt_trans_df_edge_msv$pt2)){
#     pt_trans_df_edge_msv$snv[pt_trans_df_edge_msv$old_pt1==i & pt_trans_df_edge_msv$pt2==j] <- dist_mat[i,j]
#   }
# }

pt_trans_df_edge_msv <- pt_trans_df_edge_msv %>%
  mutate(CL = ifelse(HCF1==HCF2 & CL!="Uncertain", "intra",CL))

ggplot(pt_trans_df_edge_msv) + geom_histogram(aes(x = snv)) +facet_grid(rows = vars(CL))

ggplot(pt_trans_df_edge_msv)+ geom_point(aes(x = snv, y =HCF1),color = "#F4C434") + geom_point2(aes(x = snv, y = HCF2),color = "#9ACC8F") + geom_segment(aes(y=HCF1,yend = HCF2, x = snv, xend = snv, color = CL), alpha =.3)

sum_HCF_linkage <- trans_edge_test3 %>%  # percentage of times patients were assigned to only one HCF
  group_by(n_id) %>%
  mutate(rows = row_number()) %>%
  filter(rows==1)

kruskal.test(snv~CL, data = pt_trans_df_edge_msv)

################done ##########

ggplot(sum_HCF_linkage, aes(x= freq*100, fill = as.factor(epi_type))) + geom_bar() +labs(fill ="transmission type") +scale_fill_manual( values = c("1"="#F4C434","2"="#9ACC8F","3"="#AF4474","4"="skyblue","5"="#2D4B65"), labels = c("1"="intra-facility","2"="inter-facility (acquirer)", "3"="inter-facility (source)", "4"="inter-facility (shared)", "5"="inter-facility (unknown)")) + theme_bw() + theme(panel.grid = element_blank()) + geom_vline(xintercept = 50,linetype =2, color ="red" )
#ggsave("~/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/CL_dist_hist.pdf", width = 7, height = 5)
```


```{r, fig.show='hide'}
## subset data to only keep earlies HCF link for those with common HCF links & exclude uncertain linkages  
trans_edge_test4 <- trans_edge_test3  %>%
  filter(CL!="Uncertain")%>%
  ungroup() %>%
  group_by(HCF1,HCF2, pt2, CL) %>% # finding the same healthcare facility where NDM transmitted to the same  pt2 
  arrange(pt1_date) %>% # sort by the dates of the source so the earliest one will come first
  mutate(shared_HCF = row_number(), 
         CL = ifelse(HCF1==HCF2,"intra",CL),
         to_delete = ifelse( shared_HCF!=1,1,0)) %>% ## only keeping first observed tranmission event  
  filter(to_delete==0) %>% 
  ungroup()## filter out Uncertain edges 

## check for the HCFs that are repeated or low confidence
HCF_nodes_to_delete <- setdiff(trans_edge_test3$pt1[trans_edge_test3$epi_type==4], trans_edge_test4$pt1[trans_edge_test4$epi_type==4]) 

# filter out HCFs that is low confidence from nodes list
pt_id_df2 <- pt_id_df1 %>%
  filter(pt%in%unique(c(trans_edge_test4$from,trans_edge_test4$to)))

pt_id_df2 <- pt_id_df1 %>%
  filter(pt%in%unique(c(trans_edge_test4$from,trans_edge_test4$to)))
# pt_id_df2 <- pt_id_df1 %>%
#   filter(!(pt%in%HCF_nodes_to_delete))
pt_id_df3 <- merge(pt_id_df2, patient_type,by.x = "pt", by.y = "NCBI_SAMN",all.x = TRUE)

#pt_net_test4 <- tbl_graph(pt_id_df2, trans_edge_test4 , directed = TRUE) #%>% activate(nodes) # create graph
pt_net_test4 <- tbl_graph(pt_id_df2, trans_edge_test4 , directed = TRUE) # create graph

edge_list <- pt_net_test4  %>%activate(edges) %>%as.data.frame()
node_list <- pt_net_test4  %>% activate(nodes) %>%as.data.frame()
detach
pt_net_test4 %>% ggraph(x=x, y=y) + geom_edge_fan(aes(edge_colour = CL), arrow = arrow(length = unit(1.5, 'mm'), type = "closed"), strength = 1.5,end_cap = circle(1, 'mm'), alpha =0.6,width=1) + geom_node_point(size = 3, aes(color = HCF, shape=type),alpha=.8) + scale_edge_color_manual(name = "transfer type", values = c("Genomic & Epi"="navy","Genomic only"="skyblue","intra"="#F4C434","Uncertain" = "lightblue"))+theme_graph( my_layout)+ theme(legend.position = "none") +scale_color_manual(values = location_col)+ scale_shape_manual(values = c(17, 19)) 
#ggsave("/Users/tiffanywan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/supplementary/2023_04_12_pt_network.jpg", width = 12, height = 8)
#ggplot(sum_HCF_linkage, aes(x= freq, fill = as.factor(epi_type))) + geom_bar()

#ggsave("/Users/tiffanywan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/updated_figures/pt_network_wo_legend.pdf")
```

```{r}
#######################
###   epicurve      ###
#######################
# Keep variable: sample Sample ID, sample collection HCf, sample collection date and Patient ID
iso_location0 <- patient_loc_key_ID %>%
  select(NCBI_SAMN, HCF_exposure_1, Specimen_collection_date, Patient_ID)

location_col_df <-  merge(iso_location0, location_col_df, by.x = "HCF_exposure_1",by.y = "HCF", all =TRUE)
# filter data to keep one sample per patient
iso_location1 <- iso_location0 %>%
  group_by(Patient_ID)%>%
  arrange(Specimen_collection_date)%>%
  mutate(row_num = row_number())%>%
  filter(row_num==1) # only keep one isolate per patient

# merge the HCF info with isolate collection HCF data
iso_location<- merge(iso_location1,HCF_info, by.x="HCF_exposure_1", by.y= "HCF_name", all.x = TRUE) 

# rename missing HCF type HCF
iso_location[which(is.na(iso_location$HCF_type)),"HCF_type"] <- "Veteran Affairs Hospital" 

# get the unique hospitals
iso_loc_sum <- iso_location %>%
  group_by(HCF_exposure_1, HCF_type) %>%
  summarise(n = n())
table(iso_location$HCF_type)

# create the anonymous labels
iso_loc_sum <- iso_loc_sum %>%
  group_by(HCF_type) %>%
  mutate(num = row_number(), # the numerator
         new_names = paste(HCF_type,num)) # the labels

# merge back again the iso_location  
iso_location1 <- merge(iso_location0, iso_loc_sum, by = "HCF_exposure_1") %>%
  group_by(HCF_exposure_1) %>%
  mutate(id = n(),
         inter_names = new_names,
         new_names = ifelse(id <3, "Others",new_names))

location_col_df <- merge(location_col_df, iso_location1[c("HCF_exposure_1","new_names"),], by = "HCF_exposure_1")

# set location_col vector to data frame
location_col_df <- data.frame(HCF_name = names(location_col),
                              color = location_col)

iso_location1 <- merge(iso_location1, location_col_df, by.x = "HCF_exposure_1", by.y = "HCF_name", all.x = TRUE)

iso_location1 <- iso_location1 %>%
  mutate(color=ifelse(new_names=="Others", "gray50",color))

## subset location data to new_names column and the corresponding colors
new_names_HCF_df <- iso_location1 %>%
  select(new_names,color) %>%
  distinct
```


```{r}
#  alias <- iso_location1 %>%
#    mutate(new_names1 = paste(HCF_type, num)) %>%
#    unique()
# 
# alias_v <- alias$new_names1
# names(alias_v) <- trimws(alias$HCF_exposure_1)
# 

iso_location2 <- iso_location %>%
  group_by(HCF_exposure_1) %>%
  mutate(iso_location_name = ifelse(n()<3,"Others",HCF_exposure_1)) %>%
  ungroup() %>%
  select(HCF_exposure_1,iso_location_name)

location_col_df1 <- location_col_df %>%
  merge(HCF_type, by = "HCF_name", all = TRUE) %>%
  mutate(iso = ifelse(HCF_name%in%unique(iso_location2$HCF_exposure_1),"yes","no")) %>%
  merge(iso_location2, by.x = "HCF_name",by.y = "HCF_exposure_1",all = TRUE) %>%
  distinct() %>%
  group_by(HCF_type) %>%
  mutate(pseudo_HCF = paste(HCF_type, seq_along(HCF_type))) 

location_col_df1<- location_col_df1 %>%
  filter(!is.na(HCF_type))
         
```

> Source and case curve overtime
Figure 5

```{r, fig.height=8, fig.width=9}
source_trans <- trans_edge_test3 %>%
  distinct(pt2,epi_type,CL, .keep_all = TRUE) %>%
  mutate(HCF=HCF1,
         pt=gsub("_\\d+$","",pt1),
         date =as.Date(pt2_date, origin = "1970-01-01", format = "%Y-%m-%d") ,
         Facility = "source") %>%
  ungroup() %>%
  select(pt, HCF, date, Facility)

unique(trans_edge_test3$pt2)
# method 1
# case_trans <- trans_edge_test3 %>%
#   mutate(HCF=HCF2,
#          pt=pt2,
#          date =pt2_date,
#          Facility = "case") %>%
#   select(-c(HCF1, HCF2,pt1,pt2,pt1_date,pt2_date,from,to))

# method 2: using the actual case counts and collection dates- use patient_loc_key
case_trans <- patient_loc_key_ID %>%
  mutate(pt = NCBI_SAMN,
         HCF = HCF_exposure_1,
         date = Specimen_collection_date,
         Facility = "case") %>%
  select(pt, HCF, date, Facility) %>%
  filter(pt%in%first_pt_loc$NCBI_SAMN)


trans_all <- rbind(source_trans, case_trans)
trans_all_sum <- trans_all %>%
  ungroup() %>%
  #distinct(HCF,pt, date,Facility) %>%
  group_by(HCF, Facility) %>%
  arrange(date) %>%
  mutate(x = 1,
         sum = cumsum(x)) %>%
  group_by(date, Facility, HCF) %>%
  filter(Facility =="case"|sum ==min(sum))

trans_all_sum$HCF<- trimws(trans_all_sum$HCF)
trans_all_sum_count <- trans_all_sum %>%
  mutate(var = ifelse(Facility=="source",-1,1)) %>%
  group_by(pt, HCF, date) %>%
  summarise(status = sum(var))

# summarising the number of cases/sources at one time point
 source_count <- trans_all %>%
   ungroup() %>%
   group_by(HCF, pt, date,Facility) %>%
   summarise(n=n())
 
trans_all_sum <- merge(trans_all_sum,HCF_type,by.x = "HCF",by.y = "HCF_name",all.x =TRUE)
trans_all_sum <- merge(trans_all_sum, patient_type,by.x = "pt",by.y = "NCBI_SAMN",all.x = TRUE )
trans_all_sum <- merge(trans_all_sum, source_count, by = c("HCF", "date", "Facility","pt"), all.x = TRUE )
#trans_all_sum$date <- as.Date(trans_all_sum$date, origin = "1970-01-01", format = "%Y-%m-%d")
trans_all_sum <- trans_all_sum %>%
  mutate(`Isolate Type`=ifelse(is.na(`Isolate Type`),"HCF exposure",`Isolate Type`),
         HCF_type2 = case_when(HCF_type=="Acute care hospital"~"Acute care hospital",
                               is.na(HCF_type)~"Other",
                               TRUE ~ "Other"))
## source and case curve
trans_all_sum %>%
 #filter(HCF!= "DMC Detroit Receiving Hospital") %>%
  ggplot(aes(x = date,y =sum))+geom_step(aes(group = interaction(HCF,Facility),color = HCF,linetype=Facility),alpha=0.7) +theme_bw() +theme(legend.position = "bottom", axis.text.x = element_text(angle=45, hjust=1), panel.grid.minor = element_blank()) + geom_point(aes(color = HCF,shape = `Isolate Type` ),size=2,alpha=.9) +scale_color_manual(values =setNames(location_col_df1$color, location_col_df1$HCF_name), labels = setNames(location_col_df1$pseudo_HCF, location_col_df1$HCF_name) )  + facet_grid(rows = vars(HCF_type2), scales = "free_y", space = "free_y") +scale_x_date(date_labels ="%Y-%m-%d",breaks = "3 month",limits = c(min(trans_all_sum$date), max(trans_all_sum$date))) + scale_size(range = c(1, 5)) +coord_trans( y="log2") + labs(x = "Date", y= "Number of case/source assignments to facility")+ scale_y_continuous(n.breaks = 10) + scale_shape_manual(values = c(19,15,17)) + guides(color = guide_legend(ncol = 4)) + theme(legend.box = "vertical")
ggsave("/Users/tifwan/University of Michigan Dropbox/Tiffany Wan/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/nature_communication_submission/manuscript_figures/2025_01_30_figures/fig_5.pdf",width = 9,height =11)
#ggsave("/Users/tifwan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/2024_09_09_stairs.pdf")

```

```{r}
for (i in unique(trans_all_sum$HCF)){
  df <- trans_all_sum %>%
    filter(HCF==i) 
  facility_type <- df$Facility[which(df$date==min(df$date))] 
  for (f in facility_type){
    if(f=="source"){
      print(i)
    }
  }
}
```

> Patient level transmission prediction over time
Figure 4a

```{r, fig.height=7, fig.width=12}
trans_edge_test4 <- trans_edge_test3  %>%
#  mutate(CL =ifelse(HCF1==HCF2& epi_type==1,"intra",CL)) %>% # keep all putative intra-facility transmissions even they have unknown isolates
  filter(CL!="Uncertain")%>%
  ungroup() %>%
  group_by(HCF1,HCF2, pt2, CL) %>% # finding the same healthcare facility where NDM transmitted to the same  pt2 
  arrange(pt1_date) %>% # sort by the dates of the source so the earliest one will come first
  mutate(shared_HCF = row_number(),
         CL2 = ifelse(HCF1==HCF2, "intra",CL),
         to_delete = ifelse( shared_HCF!=1,1,0)) %>% ## only keeping first observed tranmission event  
  filter(to_delete==0) %>% 
  ungroup()

## check for the HCFs that are repeated or low confidence
HCF_nodes_to_delete <- setdiff(trans_edge_test3$pt1[trans_edge_test3$epi_type==4], trans_edge_test4$pt1[trans_edge_test4$epi_type==4]) 

# filter out HCFs that is low confidence from nodes list
pt_id_df2 <- pt_id_df1 %>%
  filter(pt%in%unique(c(trans_edge_test4$from,trans_edge_test4$to)))
# pt_id_df2 <- pt_id_df1 %>%
#   filter(!(pt%in%HCF_nodes_to_delete))
pt_id_df3 <- merge(pt_id_df2, patient_type,by.x = "pt", by.y = "NCBI_SAMN",all.x = TRUE)

#pt_id_df2<- merge(pt_id_df2, alias, by.x = "HCF", by.y = "HCF_exposure_1", all =TRUE)

trans_edge_test4$pt1_date <- as.Date(trans_edge_test4$pt1_date, origin = "1970-01-01", format = "%Y-%m-%d")
trans_edge_test4$pt2_date <- as.Date(trans_edge_test4$pt2_date, origin = "1970-01-01", format = "%Y-%m-%d")
pt_id_df2$CollectionDate <- as.Date(pt_id_df2$CollectionDate, origin = "1970-01-01", format = "%Y-%m-%d")
pt_id_df3$CollectionDate <- as.Date(pt_id_df3$CollectionDate, origin = "1970-01-01", format = "%Y-%m-%d")
pt_id_df1$CollectionDate <- as.Date(pt_id_df1$CollectionDate, origin = "1970-01-01", format = "%Y-%m-%d")

# ## plot the MDHHS sample location network
# pt_trans <- pt_id_df2 %>%
#   #mutate(CL2 = ifelse(CL=="intra","Genomic & Epi"))
#   ggplot(aes(x = CollectionDate, y = HCF)) +
#   geom_curve(data = trans_edge_test4, aes(x = pt1_date, y = HCF1, xend =pt2_date, yend = HCF2, color = CL, arrow.fill = CL), arrow = arrow(length = unit(0.02,"npc"),type = "closed"),curvature = 0, ncp =8, alpha = .8) +
#   guides(color = guide_legend(ncol=3)) +
#   scale_y_discrete(limits = c("Ascension Macomb Oakland - Madison Heights Campus",
#                               "Ascension Providence Park Southfield",
#                               "Select Specialty Hospital Grosse Pointe",
#                               "Karmanos Cancer Center",
#                               "Henry Ford Hospital",
#                               "Lake Huron Medical Center",
#                               "Riverview Health & Rehab Center Jefferson",
#                               "Beaumont Royal Oak",
#                               "DMC Sinai Grace Hospital",
#                               "Ascension St. John",
#                               "Vibra DMC Campus",
#                               "Select Specialty Hospital Northwest Detroit",
#                               "Regency at Chene",
#                               "Select Specialty Hospital Downriver",
#                               "Michigan Medicine",
#                               "McLaren Port Huron",
#                               "DMC Harper University Hospital",
#                               "DMC Detroit Receiving Hospital",
#                               "Villa at Rose City",
#                               "DMC Rehabilitation Institute of Michigan",
#                               "Beaumont Taylor",
#                               "John D. Dingell Department of Veterans Affairs Medical Center",
#                               "Divinity Home Care",
#                               "Ascension Providence Novi",
#                               "Ascension Macomb Oakland - Warren Campus",
#                               "Father Murray, A Villa Center"), labels = c(alias_v,"Father Murray, A Villa Center"="Skilled nursing facility 4","Henry Ford Hospital" = "Acute care hospital 14","Regency at Chene"="Skilled nursing facility 3","Ascension Macomb Oakland - Madison Heights Campus"="Acute care hospital 13")) + scale_color_manual("Transmission type",values = c("intra"="#F4C434","Genomic only"="skyblue","Genomic & Epi"="navy"),na.value="gray", labels = c("intra"="intra-facility")) + guides(shape = guide_legend(ncol=2), fill = FALSE)+ new_scale_color() +
#   geom_point(size =3, aes(color = HCF, shape = type), alpha = 0.5 )  +scale_x_date(date_labels ="%Y-%m-%d",date_breaks = "4 months") + theme_clean() + theme(axis.text.x = element_text(angle=45, hjust=1),
#                           axis.text.y = element_text(size =7),
#                           legend.position = "bottom") + scale_shape_manual(values = c(15,19), labels = c("HCF" = "Healthcare facility", "patient"= "Patient"))  +scale_color_manual(values =location_col ) + labs(x = "Collection Date", y = "Healthcare facility", shape = "Source type", color = "Transmission type") + guides(color = FALSE)

pt_trans <- pt_id_df2 %>%
  #mutate(CL2 = ifelse(CL=="intra","Genomic & Epi"))
  ggplot(aes(x = CollectionDate, y = HCF)) +
  geom_curve(data = trans_edge_test4, aes(x = pt1_date, y = HCF1, xend =pt2_date, yend = HCF2, color = CL, arrow.fill = CL), arrow = arrow(length = unit(0.02,"npc"),type = "closed"),curvature = 0, ncp =8, alpha = .8) +
  guides(color = guide_legend(ncol=3)) +
  scale_y_discrete(limits = c("Ascension Macomb Oakland - Madison Heights Campus",
                              "Ascension Providence Park Southfield",
                              "Select Specialty Hospital Grosse Pointe",
                              "Karmanos Cancer Center",
                              "Henry Ford Hospital",
                              "Lake Huron Medical Center",
                              "Riverview Health & Rehab Center Jefferson",
                              "Beaumont Royal Oak",
                              "DMC Sinai Grace Hospital",
                              "Ascension St. John",
                              "Vibra DMC Campus",
                              "Select Specialty Hospital Northwest Detroit",
                              "Regency at Chene",
                              "Select Specialty Hospital Downriver",
                              "Michigan Medicine",
                              "McLaren Port Huron",
                              "DMC Harper University Hospital",
                              "DMC Detroit Receiving Hospital",
                              "Villa at Rose City",
                              "DMC Rehabilitation Institute of Michigan",
                              "Beaumont Taylor",
                              "John D. Dingell Department of Veterans Affairs Medical Center",
                              "Divinity Home Care",
                              "Ascension Providence Novi",
                              "Ascension Macomb Oakland - Warren Campus",
                              "Father Murray, A Villa Center"), labels = setNames(location_col_df1$pseudo_HCF, location_col_df1$HCF_name)) + scale_color_manual("Transmission type",values = c("intra"="#F4C434","Genomic only"="skyblue","Genomic & Epi"="navy"),na.value="gray", labels = c("intra"="intra-facility")) + guides(shape = guide_legend(ncol=2))+ new_scale_color() +
  geom_point(size =3, aes(color = HCF, shape = type), alpha = 0.5 )  +scale_x_date(date_labels ="%Y-%m-%d",date_breaks = "4 months") + theme_clean() + theme(axis.text.x = element_text(angle=45, hjust=1),
                          axis.text.y = element_text(size =7),
                          legend.position = "bottom") + scale_shape_manual(values = c(15,19), labels = c("HCF" = "Healthcare facility", "patient"= "Patient"))  +scale_color_manual(values = setNames(location_col_df1$color, location_col_df1$HCF_name) ,labels = setNames( location_col_df1$pseudo_HCF,location_col_df1$HCF_name)) + labs(x = "Collection Date", y = "Healthcare facility", shape = "Source type", color = "Transmission type") + guides(
    color = guide_legend(order = 1), 
    shape = guide_legend(order = 2)
  ) +
  theme(
    legend.position = "bottom",   # Place the legend on top
    legend.box = "vertical",   # Stack the legends vertically
    legend.box.spacing = unit(0.1, "cm"), # Space between the boxes
    legend.spacing.y = unit(0.1, "cm"), # Space between color/shape and fill
    legend.justification = c("left", "top"), # Position legends to top-left
    legend.key.height = unit(.3, "cm"), # Adjust size of legend items
    legend.key.width = unit(.3, "cm") # Adjust size of legend items
  ) + guides(color = FALSE)

pt_trans
ggsave("/Users/tifwan/University of Michigan Dropbox/Tiffany Wan/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/nature_communication_submission/manuscript_figures/2025_01_30_figures/fig_4a_no legend.pdf",width =8,height =7)
#ggsave("../plot/2023_04_12_pt_network_w_legend.jpg",width =11, height =7)
#ggsave("/Users/tifwan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/2024_09_09_pt_network_w_legend_alias2.pdf",width =10, height =7)
```


```{r, fig.height=7, fig.width=12}
######1. Plot the patient's MDHHS transfers first #######
# format patient_loc_ID_long1 to transfer data format (HCF->HCF2)

patient_loc_ID_long1$HCF2 <- 0
patient_loc_ID_long1 <- patient_loc_ID_long1 %>%
  arrange(Patient_ID, desc(variable)) %>% # sort by Admission date Discharge (variable column is number 1-7, from after transposing data wide to long)
  group_by(Patient_ID) %>%
  mutate(num = row_number())

```

```{r, fig.height=7, fig.width=12}
# # create HCF2 (destination)
patient_loc_ID_long1 <- patient_loc_ID_long1 %>%
  arrange(Patient_ID,NCBI_SAMN, desc(variable)) %>%
  group_by(NCBI_SAMN) %>% # transfer info per sample
  mutate(HCF2 = lead(HCF))

patient_loc_ID_long3 <- patient_loc_ID_long1 %>%
  filter(Isolate_ID%in%patient_loc$Isolate_ID) # only keep one set of transfers per patient

 # filter out transfers with no destination
patient_loc_ID_long2 <- patient_loc_ID_long3 %>%
   filter(!is.na(HCF2))
```

```{r}
patient_loc_ID_long2$HCF <- trimws(patient_loc_ID_long2$HCF) # trim trailing white space
patient_loc_ID_long2$HCF2 <- trimws(patient_loc_ID_long2$HCF2) # trim trailing white space
pt_flow_network <- patient_loc_ID_long2 %>%
  ungroup() %>%
  dplyr::select(HCF,HCF2) %>%
  group_by(HCF, HCF2) %>%
  dplyr::summarise(trans = n())

pt_flow_network <- merge(pt_flow_network, HCF_coord, by.x="HCF", by.y = "HCF_name", all.x = T )
pt_flow_network <- pt_flow_network %>%
  mutate(lon1 = lon, # rename lon
         lat1 = lat) %>% # rename lat
  dplyr::select(-c(lon, lat)) # delete the original ones

pt_flow_network<- merge(pt_flow_network, HCF_coord, by.x = "HCF2", by.y = "HCF_name", all.x = T) # merge by HCF2
pt_flow_network <- pt_flow_network %>%
  mutate(lon2 = lon,
         lat2 = lat) %>%
  dplyr::select(-c(lon,lat))

case_n <- patient_loc_key_ID_long %>%
  filter(!is.na(CollectionDate)) %>%
  select(HCF) %>%
  group_by(HCF) %>%
  summarise(case = n())

case_n$HCF <- trimws(case_n$HCF) # trim trailing white space
case_n_coord <- merge(case_n, HCF_coord, by.x="HCF", by.y = "HCF_name", all = T)
case_n_coord$case <- ifelse(is.na(case_n_coord$case),0,case_n_coord$case)
case_n_coord <- merge(case_n_coord, HCF_type, by.x = "HCF", by.y = "HCF_name", all = T)

case_n_coord1 <- case_n_coord %>%
  filter(!(HCF %in% setdiff(case_n_coord$HCF, unique(c(pt_flow_network$HCF,pt_flow_network$HCF2)))))
```


```{r}
epi_transfer_all <- data.frame(
  epi_type = pt_trans_df$epi_type,
  HCF1 = pt_trans_df$HCF1,
  HCF2 = pt_trans_df$HCF2,
  Pt_id =pt_trans_df$pt2
)
```


```{r look at all transfer groups (inter- and intra- facility)}
epi_trans_all_sum <- epi_transfer_all %>%
  group_by(HCF1, HCF2)%>%
  summarise(trans = n())

epi_trans_all_sum1<- merge(epi_trans_all_sum, HCF_coord, by.x = "HCF1", by.y = "HCF_name", all.x = T) # merge by HCF1
epi_trans_all_sum1 <- epi_trans_all_sum1 %>%
  mutate(lon1 = lon, # rename lon
         lat1 = lat) %>% # rename lat
  select(-c(lon, lat)) # delete the original ones

epi_trans_all_sum1<- merge(epi_trans_all_sum1, HCF_coord, by.x = "HCF2", by.y = "HCF_name", all.x = T) # merge by HCF2
epi_trans_all_sum1 <- epi_trans_all_sum1 %>%
  mutate(lon2 = lon,
         lat2 = lat) %>%
  select(-c(lon,lat))
```


```{r, fig.height=6, fig.width=9}
epi_trans_all_sum2 <- epi_transfer_all %>%
  group_by(HCF1, HCF2, epi_type)%>%
  summarise(trans = n())

epi_trans_all_sum2<- merge(epi_trans_all_sum2, HCF_coord, by.x = "HCF1", by.y = "HCF_name", all.x = T) # merge by HCF1
epi_trans_all_sum2 <- epi_trans_all_sum2 %>%
  mutate(lon1 = lon, # rename lon
         lat1 = lat) %>% # rename lat
  select(-c(lon, lat)) # delete the original ones

epi_trans_all_sum2<- merge(epi_trans_all_sum2, HCF_coord, by.x = "HCF2", by.y = "HCF_name", all.x = T) # merge by HCF2
epi_trans_all_sum2 <- epi_trans_all_sum2 %>%
  mutate(lon2 = lon,
         lat2 = lat) %>%
  select(-c(lon,lat))

```

> Healthcare lavel predicted transmission network
Figure 4b

```{r, fig.keep = 'first'}
case_n <- patient_loc_key_ID_long %>%
  filter(!is.na(CollectionDate)) %>%
  select(HCF) %>%
  group_by(HCF) %>%
  summarise(case = n())
case_n$HCF <- trimws(case_n$HCF) # trim trailing white space
case_n_coord <- merge(case_n, HCF_coord, by.x="HCF", by.y = "HCF_name", all = T)
case_n_coord$case <- ifelse(is.na(case_n_coord$case),0,case_n_coord$case)
case_n_coord <- merge(case_n_coord, HCF_type, by.x = "HCF", by.y = "HCF_name", all = T)

case_n_coord1 <- case_n_coord %>%
  filter(!(HCF %in% setdiff(case_n_coord$HCF, unique(c(pt_flow_network$HCF,pt_flow_network$HCF2)))))

case_n_coord2 <- case_n_coord %>% 
  filter(HCF%in%unique(c(epi_trans_all_sum2$HCF1, epi_trans_all_sum2$HCF2))) 
case_n_coord2 <- case_n_coord2 %>%
  mutate(case_b = ifelse(case==0,0,1), # boolian case variable
         id = seq(1:nrow(case_n_coord2))) %>%
  select(HCF, case, case_b,id,lon,lat) %>% # select if and the cases var
  filter(HCF%in%unique(c(epi_trans_all_sum2$HCF1, epi_trans_all_sum2$HCF2))) %>%
  mutate(x= lon,
         y= lat)

case_n_coord2$HCF <- trimws(case_n_coord2$HCF)

# merge HCF type to HCF coord df 
case_n_coord2 <- merge(case_n_coord2, HCF_type, by.x = "HCF",by.y= "HCF_name", all.x = TRUE)
#case_n_coord3 <- case_n_coord2 %>%
epi_trans_all_sum3 <- epi_trans_all_sum2 %>%
  mutate(from = HCF1,
         to = HCF2)

## create summarise plot for each HCF linkages for respective patients
summary_network <- trans_edge_test3 %>%
  distinct(HCF1, HCF2, pt2, CL) %>% # only keep distinct HCF links for each patient
  group_by(HCF1, HCF2, CL)%>% ## summarise by HCF linkage and the CL groups
  summarise(trans = n())
summary_network$HCF1 <- trimws(summary_network$HCF1)
summary_network$HCF2 <- trimws(summary_network$HCF2)

summary_network_graph <- tbl_graph(case_n_coord2, summary_network, directed = TRUE) %>% activate(nodes)
lo <- layout.fruchterman.reingold(summary_network_graph)
#detach(package:seqinr)
node_list <- summary_network_graph %>%activate(nodes)%>% as.data.frame()
summary_network_graph %>% ggraph() + 
  geom_edge_fan(aes(edge_width = as.numeric(trans), edge_colour = CL),
                strength = 1.5,
                edge_alpha = 0.7,
                arrow = arrow(length = unit(2, 'mm'),type = "closed"),
                end_cap = circle(r=3, 'mm')
                ) + 
  geom_node_point(aes(shape= as.factor(case_b), size = case, color = HCF_type),alpha=0.9) +
  theme_blank()  + 
  scale_shape_manual(values = c("0"=18, "1"=19), labels =c("No","Yes")) + 
  scale_edge_color_manual(values = c("Genomic & Epi"="#000066","Uncertain"="#c7c7ff","Genomic only"="skyblue","4"="skyblue","5"="#2D4B65"), labels = c("Uncertain" = "multiple possible\n source facilities"))+ 
  scale_size(range = c(3,12)) +scale_edge_width(range = c(1, 4.5))  + 
  labs(edge_width="# Transmissions" ,shape = "Cases >0", size = "# Cases", edge_color = "Confidence level", color = "Healthcare type") +
  scale_color_viridis_d() + theme(legend.position = "bottom")  + guides(size = "none", edge_width = "none",shape = "none")

ggsave2("~/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/2024_09_09_HCF_network1.pdf",width = 13, height = 10 )

 #ggsave("/Users/tiffanywan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/updated_figures2/HCF_network1.pdf",width = 11, height = 7.5 )

# with label
summary_network_graph %>% ggraph() + geom_edge_fan(aes(edge_width = trans, edge_colour = CL),strength = 1.5, edge_alpha = 0.4,arrow = arrow(length = unit(1.5, 'mm'),type = "closed"), end_cap = circle(1, 'mm')) + geom_node_point(aes(shape= as.factor(case_b), size = case)) +
  theme_graph() + scale_edge_alpha_continuous(range= c(0.4,1)) + scale_shape_manual(values = c("0"=18, "1"=19), labels =c("No","Yes")) + scale_edge_color_manual(values = c("Genomic & Epi"="#000066","Uncertain"="#c7c7ff","Genomic only"="skyblue","4"="skyblue","5"="#2D4B65"))+ scale_size(range = c(2.5,10)) +scale_edge_width(range = c(0.5, 4))  +geom_node_text(aes(label = HCF), size = 2,nudge_y = .1,repel = TRUE) + labs(edge_width="# Transfers" ,shape = "Cases >0", size = "# Cases", edge_color = "Confidence level") 
#ggsave("../../../../Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/updated_figures/",width = 8, height = 7 )


```


```{r}
table(patient_loc_ID_long1[!is.na(patient_loc_ID_long1$CollectionDate),]$Isolate_type)

pt_iso <- patient_loc_ID_long1[!is.na(patient_loc_ID_long1$CollectionDate),]
pt_iso_sum <- pt_iso %>%
  group_by(HCF, Isolate_type) %>%
  summarise(n = n()) %>%
  mutate(sum = sum(n),
         perc = n/sum)
  
# isolate type breakdown for isolates selected f transmission modeling
table(first_pt_loc_long[!is.na(first_pt_loc_long$CollectionDate),]$Isolate_type)
```


```{r}
pt_collection_date <- select(pt_iso, NCBI_SAMN, CollectionDate) # subset to only collection date

# merge to get date info
trans_edge_test5 <- merge(trans_edge_test3, pt_collection_date, by.x = "pt2", by.y = "NCBI_SAMN",all.x=TRUE)
## create summarise plot for each HCF linkages for respective patients
summary_network <- trans_edge_test5 %>%
  distinct(HCF1, HCF2, pt2, CL) %>% # only keep distinct HCF links for each patient
  group_by(HCF1, HCF2, CL)%>% ## summarise by HCF linkage and the CL groups
  summarise(trans = n())
```

```{r }
# add weights to the summary network
summary_network_weight  <- summary_network  %>%
  mutate(weight = trans) # add weights to network 


summary_network_weight_g <- tbl_graph(case_n_coord2, summary_network_weight, directed = TRUE) # converting nodes and edges to graph object

# change edge weights
summary_network_no_intra <- trans_edge_test3 %>%
  distinct(HCF1, HCF2, pt2, CL) %>% # only keep distinct HCF links for each patient
  group_by(pt2) %>%
  mutate(prop = 1/n()) %>% # 1/the numbers transmissions that involved given patient
  group_by(HCF1, HCF2)%>% ## summarise by HCF linkage and the groups
  mutate(trans = n()) %>% 
  filter(HCF1!=HCF2) %>% # discard self-loops
  mutate(weight = prop*trans)  %>% # add weights
  group_by(HCF1, HCF2) %>%
  summarise(weight = sum(weight)) # weight as the sum of all the weights with the same transmissions

# build graph object: case_n_coord2-nodes; summary_network_no_intra - edges
summary_network_no_intra_weight_g <- tbl_graph(case_n_coord2, summary_network_no_intra, directed = TRUE) 


eigen_inter <- eigen_centrality(summary_network_no_intra_weight_g)

# damping is arbitrary here because I don't quite get how it works
pagerank_inter <- page_rank(summary_network_no_intra_weight_g,personalized = c(ifelse(case_n_coord2$case==0, 1/sum(case_n_coord2$case),case_n_coord2$case/sum(case_n_coord2$case))), damping = 0.5) # nodes probability - proportion of cases to the total number of cases, if no cases in inferred listed HCF, assign as 1/the total number of cases; 

# combine all results
pr_df <- data.frame(HCF = case_n_coord2$HCF,
                   # rank_all = pagerank$vector,
                    rank_eigen_inter = eigen_inter$vector,
                    rank_pagerank_inter = pagerank_inter$vector)
#summary_network_no_intra_weight_g %>% tbl_df()%>%print(n=75)
```


```{r, options(dplyr.summarise.inform = FALSE), fig.show='hide'}
# get real-time eigen and pagerank ranking

# 1:  sort data by date 
trans_edge_test5 <- trans_edge_test5 %>%
  arrange(CollectionDate) 

summary_network <- trans_edge_test5 %>%
  distinct(HCF1, HCF2, pt2, CL, .keep_all = TRUE) %>% # only keep distinct HCF links for each patient
  group_by(HCF1, HCF2, CL)%>% ## summarise by HCF linkage and the CL groups
  summarise(trans = n())


rank_df <- data.frame(HCF = NULL,
                   rank_eigen = NULL,
                   rank_page = NULL,
                   rank_degree_in  = NULL,
                   rank_degree_out = NULL,
                   rank_degree_all = NULL,
                   date = NULL)

for(i in 2:nrow(trans_edge_test5)){
  date_i <- trans_edge_test5$CollectionDate[i] # Collection get the collection dates
  
  # subset edge data
  edge_df <- trans_edge_test5 %>%
  filter(CollectionDate<=date_i)
  
  ## create summarise df of each HCF linkages for respective patients
  sum_edge_df <- edge_df %>%
  distinct(HCF1, HCF2, pt2, CL) %>% # only keep distinct HCF links for each destination patient
  group_by(pt2) %>%
  mutate(prop = 1/n()) %>% 
  group_by(HCF1, HCF2, CL)%>% ## summarise by HCF linkage and the CL groups
  mutate(trans = n()) %>%
  filter(HCF1!=HCF2) %>% # only consider inter-facility transmissions
  mutate(weight = prop*trans) %>%
  group_by(HCF1, HCF2, weight) %>%
  summarise(weight = sum(weight))
  
  if(nrow(sum_edge_df)==0){
    next
  }
  
  else{
    # subset nodes data
  node_df <- case_n_coord2 %>%
    filter(HCF %in% unique(c(sum_edge_df$HCF1, sum_edge_df$HCF2)))
  
  # combine nodes and edges into graph object
  network_g <- tbl_graph(node_df, sum_edge_df, directed = TRUE)
  
  eigen <- eigen_centrality(network_g)
  degree_out <- igraph::degree(network_g, mode = "out")
  degree_in <- igraph::degree(network_g, mode = "in")
  degree_all <- igraph::degree(network_g, mode = "total")
  page <- page_rank(network_g,personalized = c(ifelse(node_df$case==0, 1/sum(node_df$case),node_df$case/sum(node_df$case))), damping = 0.5)

  
  df <- data.frame(HCF = node_df$HCF,
                   rank_eigen = eigen$vector,
                   rank_page = page$vector,
                   rank_degree_out = degree_out,
                   rank_degree_in = degree_in,
                   rank_degree_all = degree_all,
                   date = date_i)
  
  }
  rank_df <- rbind(rank_df, df) # bind new rankings with existing rankings

}


rank_df <- rank_df %>%
  mutate(rank_eigen=round(rank_eigen,5)) %>%
    distinct() %>%
  group_by(date) %>%
  mutate(prop_eigen = rank_eigen/sum(rank_eigen),
         prop_page = rank_page/sum(rank_page),
         prop_degree_in = rank_degree_in/sum(rank_degree_in),
         prop_degree_out = rank_degree_out/sum(rank_degree_out),
         prop_degree_all = rank_degree_all/sum(rank_degree_all))

full_rank <- rank_df %>%
  filter(date=="2022-05-27")

ggplot(full_rank, aes(x = rank_eigen,y = rank_page, color = HCF)) +geom_point() +  scale_color_manual(values = location_col ) +theme(legend.position = "bottom",legend.title = element_blank(),legend.text = element_text(size=10),legend.key.size = unit(1,"line")) + geom_smooth(method = "lm", se = TRUE) + guides(color=guide_legend(ncol=2))
#ggsave("../../../../Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/exploratory/compared_pagerank_eigen.pdf",width = 8, height = 7)

ggplot(full_rank, aes(x = rank_page,y = rank_degree_all, color = HCF)) +geom_point() +  scale_color_manual(values = location_col ) +theme(legend.position = "bottom",legend.title = element_blank(),legend.text = element_text(size=10),legend.key.size = unit(1,"line"))

ggplot(full_rank, aes(x = rank_eigen,y = rank_degree_all, color = HCF)) +geom_point() +  scale_color_manual(values = location_col ) +theme(legend.position = "bottom",legend.title = element_blank(),legend.text = element_text(size=10),legend.key.size = unit(1,"line")) + geom_smooth(method = "lm", se = TRUE) + guides(fill=guide_legend(ncol=2))
# eigen_rank <- full_rank %>%
#   arrange(desc(rank_page)) 
#   
# head(eigen_rank,10)

rank_df_wide <- rank_df %>%
  pivot_wider(id_cols = HCF, names_from = date, values_from = rank_eigen)

# out-degree centrality - influential 
ggplot(rank_df) +geom_bar(aes(x = as.factor(date), y = prop_degree_out, fill = reorder(HCF, prop_degree_out)),stat = "identity",position = "stack") +
  theme(legend.position = "bottom",legend.title = element_blank(),legend.text = element_text(size=10),legend.key.size = unit(1,"line"),
        axis.text.x = element_text(angle = 45,hjust=1)) + scale_fill_manual(values = location_col ) + guides(fill=guide_legend(ncol=3)) +labs(y = "Centrality Importance", x = "Date")+ guides(fill=guide_legend(ncol=2))

ggplot(rank_df) +geom_bar(aes(x = as.factor(date), y = prop_eigen, fill = reorder(HCF, prop_eigen)),stat = "identity",position = "stack") +
  theme_bw()+
  theme(legend.position = "bottom",legend.title = element_blank(),legend.text = element_text(size=10),legend.key.size = unit(1,"line"),
        axis.text.x = element_text(angle = 45,hjust=1)) + scale_fill_manual(values = location_col ) + guides(fill=guide_legend(ncol=3)) +labs(y = "Centrality Importance", x = "Date")


  #ggsave("/Users/tiffanywan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/updated_figures2/HCF_ranking_eigen.pdf", width = 12, height=8)
ggplot(rank_df) +geom_bar(aes(x = as.factor(date), y = prop_page, fill = reorder(HCF, prop_page)),stat = "identity",position = "stack") +
  theme_bw()+
  theme(legend.position = "bottom",legend.title = element_blank(),legend.text = element_text(size=10),legend.key.size = unit(1,"line"),
        axis.text.x = element_text(angle = 45,hjust=1)) + scale_fill_manual(values = location_col ) + guides(fill=guide_legend(ncol=3))+ labs(y = " Pagerank Importance", x = "Date") 


#ggsave("/Users/tiffanywan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/updated_figures2/HCF_ranking_page.pdf", width = 12, height=8)
```

```{r, fig.show='hide'}
library(plotly)
e_g <- ggplot(rank_df) +geom_line(aes(x = as.factor(date), y = rank_eigen, group = HCF, color = HCF)) + geom_point(aes(x = as.factor(date), y = rank_eigen,color = HCF), alpha = .4) +
  theme_bw()+
  theme(legend.position = "bottom",legend.title = element_blank(),legend.text = element_text(size=10),legend.key.size = unit(1,"line"),
        axis.text.x = element_text(angle = 45,hjust=1)) + scale_color_manual(values = location_col ) + guides(color=guide_legend(ncol=3))+ labs(y = " Eigen Importance", x = "Date") 
#ggsave("/Users/tifwan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/HCF_ranking_eigen_trend.pdf", width = 12, height=8)
l <- ggplotly(e_g)
htmlwidgets::saveWidget(l, "html_ouput/eigen_trend.html")

ggplot(rank_df) +geom_line(aes(x = as.factor(date), y = log10(rank_page), group = HCF, color = HCF)) + geom_point(aes(x = as.factor(date), y = log10(rank_page),color = HCF), alpha = .4) +
  theme_bw()+
  theme(legend.position = "bottom",legend.title = element_blank(),legend.text = element_text(size=10),legend.key.size = unit(1,"line"),
        axis.text.x = element_text(angle = 45,hjust=1)) + scale_color_manual(values = location_col ) + guides(color=guide_legend(ncol=3))+ labs(y = " Pagerank Importance", x = "Date") 
#ggsave("/Users/tifwan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/HCF_ranking_page_trend_log.pdf", width = 12, height=8)

```


```{r, fig.show='hide'}
# subset dataset to patient and collection date for plotting cumulative case barplot by time
cum_pt_date <- first_pt_loc %>%
  select(NCBI_SAMN, Specimen_collection_date, HCF_exposure1) %>%
  mutate(Specimen_collection_date=as.Date(Specimen_collection_date, format = "%m/%d/%y")) %>%
 # filter(Specimen_collection_date%in%rank_df$date) %>%
  distinct(Specimen_collection_date,.keep_all = TRUE) %>%
  arrange(Specimen_collection_date)
  
cum_pt_date1 <- cum_pt_date
cum_pt_date1$HCF_exposure1 <- trimws(cum_pt_date1$HCF_exposure1)

## create cumulative case barplot
HCF_v <- NULL
date_v <- NULL
for(i in 1:nrow(cum_pt_date1)){
  v1 <- rep(cum_pt_date1$Specimen_collection_date[i],i) # repeat the collection date i times
  date_v<- c(date_v, v1) # bind all the repeated collection dates together 
  v2 <- cum_pt_date1$HCF_exposure1[1:i] # the cumulative case collection HCF
  HCF_v <- c(HCF_v,v2) # bind the collection HCFs together
}

HCF_sum_date <- data.frame(
  HCF = HCF_v,
  date = date_v
)
HCF_sum_date$date <- as.Date(HCF_sum_date$date,origin ="1970-01-01")
HCF_sum_date %>%
  ggplot(aes(x = as.factor(date), fill = HCF))+geom_bar() + theme_bw()+theme(legend.position = "none", axis.text.x = element_text(angle=45,hjust = 1)) +scale_fill_manual(values = location_col) 
#ggsave("/Users/tiffanywan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/updated_figures/HCF_date_cum_bar.pdf", width = 12, height=5)
```



```{r, fig.show='hide'}
# function for aligning the tree tips to the plot y-axis
tree_y <-  function(ggtree, data){
  if(!inherits(ggtree, "ggtree"))
    stop("not a ggtree object")
  left_join(select(data, label), select(ggtree$data, label, y)) %>%
    pull(y)
}
patient_loc_ID_long1$label <- patient_loc_ID_long1$NCBI_SAMN

patient_loc_ID_long_test <- patient_loc_ID_long1 %>%
  group_by(NCBI_SAMN) %>%
  arrange(NCBI_SAMN,variable) %>%
  mutate(DischargeDate=ifelse(!is.na(CollectionDate),CollectionDate,DischargeDate))

patient_loc_ID_long_test$label <- patient_loc_ID_long_test$NCBI_SAMN
# make sure the name of the tree label.tips matches the list 
MDHHS_list1 <- MDHHS_list
MDHHS_list1[MDHHS_list1$x=="20MP010425-WI-M3453-201110_S1",2] <- "20MP010425-WI-M3453-201110"

# create tree metadata that flags patients if they have more than one isolate collected
tree_metadata <- patient_loc_key_ID %>%
  select(NCBI_SAMN,Patient_ID) %>%
  group_by(Patient_ID)%>%
  mutate(count = n(),
         id = ifelse(count==1,"none",Patient_ID)) 
# reroot tree
which(iqtree$tip.label=="SAMEA3531785")
iqtree <- reroot(iqtree, which(iqtree$tip.label=="SAMEA3531785"))
iqtree$tip.label[which(iqtree$tip.label=="SAMEA3531785")]


# keep tips
iqtree_sub <- keep.tip(iqtree, MDHHS_list1$x)
iqtree_sub <- drop.tip(iqtree_sub,"SAMN24979401") # this isolate is likely contaminated
tree_metadata1 <- tree_metadata %>%
  filter(NCBI_SAMN!="SAMN24979401") # filter out contaminated strain
iqtree_p_sub <- ggtree(iqtree_sub, aes(col = c(tree_metadata$id,rep("none",times =103)))) 

# plot iqtree results with only the MDHHS sample
#iqtree_sub <- keep.tip(iqtree, MDHHS_list1$x)
iqtree_p_sub <- ggtree(iqtree_sub) 
id_col <- c("#8fc2ef", "#a8f3f7", "#251ca8", "#8d37d3", "#edae31", "#9b9af4", "#bec6f7", "#fcfcba", "#e1ea3c", "#8831b7", "#a4f492", "#31e2a2", "#50c94a", "#e026c1", "#dbc16d", "#7d46fc", "#15912a", "#3c75ad", "#d8e582", "#633daf", "#b2fffc", "#ff7847", "#9ffcbe")
iqtree_p_sub1 <- iqtree_p_sub %<+% tree_metadata + geom_tippoint(aes(color = id),size=.8, alpha = 0.6) + scale_color_manual(values = c(id_col,"black")) +theme(legend.position = "left") +
  theme(legend.position = "none")
iqtree_p_sub1

id_col <- c("#8fc2ef", "#a8f3f7", "#251ca8", "#8d37d3", "#edae31", "#9b9af4", "#bec6f7", "#fcfcba", "#e1ea3c", "#8831b7", "#a4f492", "#31e2a2", "#50c94a", "#e026c1", "#dbc16d", "#7d46fc", "#15912a", "#3c75ad", "#d8e582", "#633daf", "#b2fffc", "#ff7847", "#9ffcbe")

# iqtree_p_sub1 <- iqtree_p_sub %<+% tree_metadata + geom_tippoint(aes(shape = id),size=1.5, alpha = 0.6)  +theme(legend.position = "left") +
#   theme(legend.position = "none")

# iqtree_p_sub %<+% tree_metadata +
#   geom_tiplab(aes(label=NA, col=c(tree_metadata$id,rep("none",times =103))), hjust=-0.5, align=TRUE, linesize=0.1)
# iqtree_p_sub1 <- iqtree_p_sub %<+% tree_metadata + geom_tippoint(aes(color = id),size=1.5) + scale_color_manual(values = c(id_col,"black")) +theme(legend.position = "left") +
#   theme(legend.position = "none")
# iqtree_p_sub1
 #ggtree(keep.tip(iqtree, first_isolate$NCBI_SAMN)) + geom_tiplab(size=2.5)
iqtree_p_sub + geom_tiplab(size=2)
```

```{r, fig.show='hide'}
p <-ggplot(patient_loc_ID_long1)+
  geom_segment(aes(x = AdmitDate, xend = DischargeDate, # add one to work around startdates that are the same as enddates
                   y = tree_y(iqtree_p_sub, patient_loc_ID_long1), yend= tree_y(iqtree_p_sub, patient_loc_ID_long1), colour = HCF),size =2.5) +
  geom_point(aes(x = CollectionDate,y = tree_y(iqtree_p_sub, patient_loc_ID_long1),colour = Isolate_type),size = 1) +
  scale_color_manual(values = c(setNames(location_col_df1$color,location_col_df1$HCF_name),"Surveillance" = "red","Clinical"="black"))+
  theme_minimal() +
  theme(panel.grid.minor.x = element_blank(),
        panel.grid.major.x =element_blank(),
        axis.title = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1),
        axis.ticks.length = unit(0, "pt"),
        plot.margin = margin(0,0,0,0),
        legend.position = "none") +
  guides(color = guide_legend(ncol = 1, byrow = TRUE)) +
  scale_x_date(date_breaks="6 months",date_labels="%b %Y") + scale_y_continuous(expand=c(0,0))

patient_loc_ID_long_test$DischargeDate <- as.Date(patient_loc_ID_long_test$DischargeDate,origin ="1970-01-01")
p_test <- ggplot(patient_loc_ID_long_test)+
  geom_segment(aes(x = AdmitDate, xend = DischargeDate, # add one to work around startdates that are the same as enddates
                   y = tree_y(iqtree_p_sub, patient_loc_ID_long1), yend= tree_y(iqtree_p_sub, patient_loc_ID_long1), colour = HCF),size =1.5) +
  geom_point(aes(x = CollectionDate,y = tree_y(iqtree_p_sub, patient_loc_ID_long1),colour = Isolate_type),size = 0.5) +
  scale_color_manual(values = c(setNames(location_col_df1$color,location_col_df1$HCF_name),"Surveillance" = "red","Clinical"="black"))+
  theme_minimal() +
  theme(axis.title = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1),
        legend.position = "none") +
  guides(color = guide_legend(ncol = 1, byrow = TRUE)) +
  scale_x_date(date_breaks="6 months",date_labels="%b %Y") 
p_test1 <- p_test + ylim2(iqtree_p_sub1,limits = c(0.5,103.5)) 

#ggsave("2023_04_12_ptmvmt_collectiondate.jpg", height = 9, width = 10)
p1 <- p + ylim2(iqtree_p_sub1,limits = c(0.5,103.5)) 

insert_left(p1, iqtree_p_sub1, width=0.4)
iqtree_p_sub1 %>% insert_right(p1)
```


```{r}
# overwrite the default expand for continuous scales
scale_y_tree <- function(expand=expand_scale(0, 0.6), ...){
    scale_y_continuous(expand=expand, ...)
}

# get the range of the ggtree y-axis data
tree_ylim <- function(ggtree){
  if(!inherits(ggtree, "ggtree"))
    stop("not a ggtree object")
  range(ggtree$data$y)
}

# plot data next to a ggtree aligned by shared labels
ggtreeplot <- function(ggtree, data = NULL, mapping = aes(), flip=FALSE,
     expand_limits=expand_scale(0,.6), ...){
  
  if(!inherits(ggtree, "ggtree"))
    stop("not a ggtree object")

  # match the tree limits
  limits <- tree_ylim(ggtree)
  limits[1] <- limits[1] + (limits[1] * expand_limits[1]) - expand_limits[2]
  limits[2] <- limits[2] + (limits[2] * expand_limits[3]) + expand_limits[4]
  
  if(flip){
    mapping <- modifyList(aes_(x=~x), mapping)
    data <- mutate(data, x=tree_y(ggtree, data))
    gg <- ggplot(data=data, mapping = mapping, ...) +
      scale_x_continuous(limits=limits, expand=c(0,0))
  }else{
    mapping <- modifyList(aes_(y=~y), mapping)
    data <- mutate(data, y=tree_y(ggtree, data))
    gg <- ggplot(data=data, mapping = mapping, ...) +
      scale_y_continuous(limits=limits, expand=c(0,0))
  }
  gg
}

# get rid of superfluous axis - this works after coord_flip, so it also works
# for the rotated histogram
no_y_axis <- function () 
  theme(axis.line.y = element_blank(), 
        axis.title.y = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank())
```

```{r}
# merge the isolate-patient match ids to the test_patient_transfer_long3 data
iso_pt_ID <- patient_loc_key_ID %>%
  select(Patient_ID, NCBI_SAMN)
```

```{r}
trans_edge_test3 <-  pt_trans_df_edge %>%
  select(epi_type, HCF1, HCF2, pt1, pt2, pt1_date, pt2_date, from, to) %>%
  group_by(pt2, HCF1, HCF2) %>%
  mutate(pt_link_n= n(), ## count the numbers of HCF links in regard to one patient
         n_id = cur_group_id(),## assign group ids to these groups
         genomic_only = ifelse(epi_type==5,1,0)) %>% 
  group_by(pt2) %>%
  mutate(n=n(), ## count the numbers of time a patient was a assigned to a group
         freq = pt_link_n/n, ## count the percentage of the times groups were being assigned per patient
         ## create arbitrary confidece level variable
         CL = case_when(epi_type!=5 & freq >0.5 ~"Genomic & Epi",
                        epi_type==5 & freq >0.5 ~"Genomic only",
                        TRUE ~ "Uncertain"))  


# Get unique HCF/ or uncertain HCF for those with only uncertain HCFs 
predicted_trans <- trans_edge_test3  %>%
  select(pt2, CL,HCF1,HCF2) %>%
  distinct(pt2,HCF1, .keep_all = TRUE) %>%
  group_by(pt2,CL)%>%
  mutate(n=n()) %>%
  filter(!(CL=="Uncertain" & n==1)) %>%
  group_by(pt2) %>%
  mutate(n=n()) %>%
  mutate(epi_type = CL,
         NCBI_SAMN = pt2) %>%
  ungroup()%>%
  mutate(CL2= ifelse(HCF1==HCF2,"intra-facility",CL)) %>%
  select(-c(CL, pt2)) 


# get the epi_type for all data
predicted_trans1 <- merge(predicted_trans, iso_pt_ID, by = "NCBI_SAMN", all.x = TRUE)
predicted_trans1 <- merge(predicted_trans1, iso_pt_ID, by = "Patient_ID", all.x = TRUE) %>%
  mutate(NCBI_SAMN = NCBI_SAMN.y) %>%
  select(-c(NCBI_SAMN.x,NCBI_SAMN.y,)) %>%
  mutate(label = NCBI_SAMN)
  
```

```{r,fig.show='hide'}
# merge data to include epi_type assignment for not just the first case per patient
MDHHS_test1 <- merge(MDHHS_test, iso_pt_ID, by.x = "Patient_ID",by.y = "Patient_ID", all.x = TRUE) %>%
  mutate(NCBI_SAMN = NCBI_SAMN.y) %>%
  select(-c(NCBI_SAMN.x,NCBI_SAMN.y,))


MDHHS_test1_update <- MDHHS_test1 %>%
  arrange(-update) %>% 
  distinct(NCBI_SAMN, source, epi_HCF, epi_type, .keep_all = TRUE) # remove the old unknowns that were updated after having epi_type==2
MDHHS_test1_update$label <- MDHHS_test1_update$NCBI_SAMN
MDHHS_test1$label <- MDHHS_test1$NCBI_SAMN 
  

MDHHS_source <- MDHHS_test1_update %>% 
  mutate(NCBI_SAMN = ifelse(update==0,NCBI_SAMN,source), # change source of updated data to NCBI_SAMN because the sources were in opposite directions
         source_HCF = from)
MDHHS_source$label <- MDHHS_source$NCBI_SAMN

# plot transfer groups 
ggplot(data = MDHHS_test1_update, aes(x = NCBI_SAMN, fill = as.factor(epi_type))) + geom_bar(stat = "count") + coord_flip() +
  scale_fill_manual(name = "transfer type", values = c("1"="#F4C434","2"="#9ACC8F","3"="#AF4474","4"="skyblue","5"="#2D4B65"), labels = c("1"="intra-facility","2"="inter-facility (acquirer)", "3"="inter-facility (source)", "4"="inter-facility (shared)", "5"="inter-facility (unknown)"))  +
  scale_y_continuous(breaks = c(0, 5, 10))
```


```{r,fig.height=8, fig.width=8,fig.show='hide'}

# get the number of unque HCFs per isolate of interest
unique_HCF<- MDHHS_source  %>%
  group_by(NCBI_SAMN, Patient_ID) %>% # keep only HCFs identified grouped by everything listed below epitype
  summarise(uniq_HCF = n_distinct(from)) %>%
  ungroup() %>%
  distinct(Patient_ID, .keep_all = TRUE)

summary(unique_HCF$uniq_HCF)
ggplot(unique_HCF, aes(x = uniq_HCF))+ geom_bar() 



```

## Transmission source predictions after using 50% threshold filtering bar plot 
>Plotting tree, trace and barplot together
Figure 2

```{r,fig.height=8, fig.width=8, fig.keep='first'}
## for the purpose of plotting the barplot let all the transfers with the same HCF source be added up to 1
MDHHS_source1 <- MDHHS_source %>%
  distinct(Patient_ID, source_HCF, epi_type, NCBI_SAMN, label) %>%# keep only HCFs identified grouped by everything listed below epitype
    group_by(NCBI_SAMN,Patient_ID) %>%
  mutate(uniq_HCF = n_distinct(source_HCF)) %>%
  group_by(NCBI_SAMN,Patient_ID, source_HCF) %>%
  mutate(count = 1/n()) # let same source_HCF with different epi_type be expressed as proportion that adds up to 1
  
# # create bar plot for testing the groupings
# bar <- ggplot(data = MDHHS_source1, aes(x = tree_y(iqtree_p_sub, MDHHS_source1), y = count, fill = as.factor(epi_type))) + geom_bar(stat = "identity") +coord_flip()+
#   theme(axis.title.y = element_blank(), # remove title
#         axis.text.y=element_blank(), # remove text
#         axis.ticks.y = element_blank(), # remove ticks
#         axis.title.x = element_text(size=10),
#         legend.position = "none") + # remove legend
#   scale_y_continuous(name = "Number of source \nFacilities",breaks = c(2,4, 6, 8,10), expand = c(0.01,0)) +
#   scale_x_continuous(expand = c(0,0)) +
# scale_fill_manual(name = "transfer type", values = c("1"="#F4C434","2"="#9ACC8F","3"="#AF4474","4"="skyblue","5"="#2D4B65"), labels = c("1"="intra-facility","2"="inter-facility (acquirer)", "3"="inter-facility (source)", "4"="inter-facility (shared)", "5"="inter-facility (unknown)")) 
#  

predicted_trans2 <- predicted_trans1 %>%
  distinct(label, .keep_all = TRUE)
# bar plot #2: this time using CL groupings
bar <- ggplot(data = predicted_trans2, aes(x = tree_y(iqtree_p_sub, predicted_trans2), y = n, fill = as.factor(epi_type))) + geom_bar(stat = "identity") +coord_flip()+ theme_bw()+
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        axis.title.y = element_blank(), # remove title
        axis.text.y=element_blank(), # remove text
        axis.ticks.y = element_blank(), # remove ticks
        axis.title.x = element_text(size=10),
        legend.position = "none") + # remove legend
  scale_y_continuous(name = "Number of source \nFacilities",breaks = c(2,4, 6, 8,10), expand = c(0,0)) +
  scale_x_continuous(expand = c(0,0)) +
scale_fill_manual(name = "transfer type", values = c("Genomic & Epi"="#000066","Uncertain"="#c7c7ff","Genomic only"="skyblue","intra-facility"="#F4C434"), labels = c("1"="intra-facility","2"="inter-facility (acquirer)", "3"="inter-facility (source)", "4"="inter-facility (shared)", "5"="inter-facility (unknown)"))  
#ggsave("~/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/bar_with legend.pdf")

#bar
iqtree_p_sub2 <- iqtree_p_sub1 +geom_tiplab(size =2)
p1 <- p + ylim2(iqtree_p_sub2,limits = c(0.5,103.5)) 

p2 <- insert_left(p1, iqtree_p_sub1, width = .45)

#iqtree_p_sub2 <-  iqtree_p_sub+ geom_tiplab(size = 2)
insert_right(p2, bar, width = .2)
ggsave("/Users/tifwan/University of Michigan Dropbox/Tiffany Wan/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/nature_communication_submission/manuscript_figures/2025_01_30_figures/fig_2.pdf", height = 9, width = 9)
#ggsave("/Users/tiffanywan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/updated_figures2/2023_10_25_phy_pt_mvmt_bar_lab.jpg", height = 12, width = 23)

#ggsave("../plot/2023_04_12_phy_pt_mvmt_bar_lab.jpg", height = , width = 12)

p_test2 <- insert_left(p_test1 , iqtree_p_sub1, width = .45)
p_test2 <- insert_left(p1, iqtree_p_sub1, width=0.4)
fig_2_no_legend <- insert_right(p_test2 , bar, width = .2)
legend_all_color <- ggplot(location_col_df1, aes(x = pseudo_HCF, fill = pseudo_HCF)) + geom_bar() + scale_fill_manual(values = setNames(location_col_df1$color,location_col_df1$pseudo_HCF)) + theme(legend.position = "bottom",               # Position legend below plot
    legend.title = element_text(size = 8),     # Legend title text size
    legend.text = element_text(size = 8),      # Legend label text size
    legend.key.size = unit(0.2, "cm"),          # Smaller legend color keys
    legend.spacing.x = unit(0.1, "cm") ) + guides(fill = guide_legend(ncol = 8))

#ggsave("/Users/tifwan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/2023_04_12_phy_pt_mvmt_bar_lab_collec_NEW.pdf", height = 9, width = 9)
#ggsave("/Users/tiffanywan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/updated_figures2/2023_04_12_phy_pt_mvmt_bar_lab_collec.pdf",height = 9, width = 10)
```

> Patient healthcare exposure and predicted source facilities line plot
Figure 3b

```{r,fig.keep='last'}
# number of transfers
patient_transfers_sum <- patient_loc_ID_long1 %>%
  distinct(Patient_ID,NCBI_SAMN, HCF) %>% # find the distinct HCF exp. per patient
  group_by(Patient_ID,NCBI_SAMN) %>% 
  summarise(n=n()) %>%# calculate the numbers of HCFs per patient
  #select(-NCBI_SAMN) %>%
  distinct(Patient_ID, .keep_all = TRUE) 
patient_transfers_sum$stat <- "transfer"

# number of of transmission groups
unique_HCF_sub <- unique_HCF %>%
  select(Patient_ID, uniq_HCF, NCBI_SAMN) %>%
  mutate(stat = "all transmission",
         n = uniq_HCF) %>%
  select(-uniq_HCF)

# final number of predicted transmissions
predicted_trans <- trans_edge_test3  %>%
  select(pt2, CL,HCF1) %>%
  distinct(pt2,HCF1, .keep_all = TRUE) %>%
  group_by(pt2, CL)%>%
  mutate(n=n()) %>%
  filter(!(CL=="Uncertain" & n==1)) %>%
  group_by(pt2,CL) %>%
  summarise(n=n()) %>%
  mutate(stat="filtered transmission") %>%
  mutate(epi_type = CL,
         NCBI_SAMN = pt2) %>%
  ungroup()%>%
  select(-c(CL, pt2))


#### Adding an additional step to filter out uncertain facilities for intra-facility transmissions  ####
predicted_trans <- trans_edge_test3  %>%
  select(pt2, CL,HCF1,HCF2) %>%
  distinct(pt2,HCF1, .keep_all = TRUE) %>%
  group_by(pt2,CL)%>%
  mutate(n=n()) %>%
  filter(!(CL=="Uncertain" & n==1)) %>%
  group_by(pt2) %>%
  mutate(n=n()) %>%
  mutate(epi_type = CL,
         NCBI_SAMN = pt2) %>%
  ungroup()%>%
  mutate(CL2= ifelse(HCF1==HCF2,"intra-facility",CL)) # find the ones that are intra-faciltiy

predicted_trans <-predicted_trans[-which(predicted_trans$CL!=predicted_trans$CL2 & predicted_trans$CL=="Uncertain"),] # find the index of the 'Uncertain-intrafacility links'

predicted_trans <- predicted_trans %>%
  group_by(pt2,CL) %>%
  summarise(n=n()) %>%
  mutate(stat="filtered transmission") %>%
  mutate(epi_type = CL,
         NCBI_SAMN = pt2) %>%
  ungroup()%>%
  select(-c(CL, pt2))
##########################  
  
compare_trans_predictions <- rbind(patient_transfers_sum,unique_HCF_sub) %>%
  # group_by(stat, n) %>%
  # mutate(edge = n())%>%
  group_by(NCBI_SAMN) %>%
  mutate(ratio = n/sum(n)) %>%
  filter(n()==2) # filter out data without transfer-transmission pair (eg: index case, horizontal gene transfer pt)

# isolates that have more transfers than there were predicted transmissions
print(paste("Isolates that have more transmission predictions than there were transfers:",length(unique(compare_trans_predictions$Patient_ID[compare_trans_predictions$ratio<.5 & compare_trans_predictions$stat=="transfer"])),"/",length(unique(compare_trans_predictions$Patient_ID))))

unique(compare_trans_predictions$Patient_ID[compare_trans_predictions$ratio<.5 & compare_trans_predictions$stat=="transfer"]) # NDM68 was epi_type==2 others were all epitype==5(unknown) 
# length = 6
print(paste("Isolates that have the same number of transmission predictions as there were transfers:",length(unique(compare_trans_predictions$Patient_ID[compare_trans_predictions$ratio==.5 & compare_trans_predictions$stat=="transfer"])),"/",length(unique(compare_trans_predictions$Patient_ID))))
unique(compare_trans_predictions$Patient_ID[compare_trans_predictions$ratio==.5 & compare_trans_predictions$stat=="transfer"]) # length = 16

print(paste("Isolates that have less transmission predictions as there were transfers:",length(unique(compare_trans_predictions$Patient_ID[compare_trans_predictions$ratio>.5 & compare_trans_predictions$stat=="transfer"])),"/",length(unique(compare_trans_predictions$Patient_ID))))
unique(compare_trans_predictions$Patient_ID[compare_trans_predictions$ratio>.5 & compare_trans_predictions$stat=="transfer"]) # length = 50


# compare_trans_predictions %>%
#   group_by(stat, Patient_ID,n) %>%
#   summarise()
compare_trans_predictions_epi <- merge(compare_trans_predictions, MDHHS_source, by.x = "NCBI_SAMN", by.y = "NCBI_SAMN" ,all.x = TRUE) %>%
  distinct(NCBI_SAMN,epi_type, n,stat) 

compare_trans_predictions_epi <- rbind(compare_trans_predictions_epi, predicted_trans) %>%
  mutate(n_jitter = jitter(n,.3))

compare_trans_predictions_epi$stat1 <- factor(compare_trans_predictions_epi$stat, levels = c("transfer","all transmission","filtered transmission"))

#stat_level <- factor(compare_trans_predictions$stat, c("transfer","transmission","filtered transmission"))
# renaming values that goes into the axis values: here we are changing variable stat to have legible values
# compare_trans_predictions_epi <- compare_trans_predictions_epi %>%
#   mutate(stat1 = ifelse(stat=="transfer", "Facilities exposed to","Predicted isolate source facilities"))
  
#prediction_sig_test <- cochran_qtest(compare_trans_predictions, n ~ stat|Patient_ID)
# here patient transmission predicted to be in two groups would be shown in two separate lines, but the number of transfer - transmission would stay the same

compare_trans_predictions_epi %>%
  #filter(epi_type==1) %>%
  ggplot( aes(x = stat1, y = n_jitter , color = as.factor(epi_type),group = NCBI_SAMN)) +geom_line(alpha = .7) +scale_color_manual(name = "transfer type", values = c("1"="#F4C434","2"="#9ACC8F","3"="#AF4474","4"="skyblue","5"="#2D4B65", "Genomic & Epi"= "navy","Genomic only"="skyblue","Uncertain"="#c7c7ff"), labels = c("1"="intra-facility","2"="inter-facility (acquirer)", "3"="inter-facility (source)", "4"="inter-facility (shared)", "5"="inter-facility (unknown)")) +scale_y_continuous(n.breaks = 9) 
#ggsave("../../../../Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/updated_figures/compare_prediction_tranfer.pdf")

compare_trans_predictions_epi %>%
  #filter(epi_type==1) %>%
  ggplot() + geom_violin(aes(x = stat1,y = n),color = NA, fill = 'navy', alpha=.3)+geom_line(aes(x = stat1, y = n_jitter ,group = NCBI_SAMN),alpha = .7) + theme_bw() +scale_y_continuous(n.breaks = 9) +labs(x = "",y = "Number of facilities") + scale_x_discrete(labels = c("Number of facilities\n acquiring patients\n exposed to","Number of predicted\n isolate source facilities","Number of predicted\n isolate source facilities\n with source filtering"))  +theme(panel.grid.minor = element_blank(), panel.grid.major.y = element_blank(),axis.text.x = element_text(size=11))
#ggsave("../../../../Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/compare_prediction_tranfer2.pdf",width = 6, height=6)

filtered_summary <- compare_trans_predictions_epi%>%
  filter(stat1 =="filtered transmission")
summary(filtered_summary$n)
```




```{r}
# t.test(compare_trans_predictions$n[compare_trans_predictions$stat=="transfer"], compare_trans_predictions$n[compare_trans_predictions$stat=="transmission"], paired = TRUE, alternative = "two.sided")
# 
# print("Statistics of predicted number of sources per patient")
# summary(unique_HCF_sub$n)
# table(unique_HCF_sub$n)
```


```{r, fig.show='hide'}
first_pt_incidence_mat <- first_pt_loc_long %>%
  group_by(NCBI_SAMN,HCF) %>%
  summarise(value=n())%>%
  pivot_wider(id_cols = NCBI_SAMN,
              names_from = HCF,
              values_from=value,
              values_fill = 0)

first_pt_incidence_mat1 <- first_pt_incidence_mat %>%
  column_to_rownames(var = "NCBI_SAMN") %>%
  as.matrix()
pheatmap(first_pt_incidence_mat1,
         border_color = NA,
         angle_col = "315",
         fontsize = 5)

all_patient_incidence_mat <- patient_loc_ID_long1 %>%
   group_by(NCBI_SAMN,HCF) %>%
  summarise(value=n())%>%
  pivot_wider(id_cols = NCBI_SAMN,
              names_from = HCF,
              values_from=value,
              values_fill = 0)

all_patient_incidence_mat1 <- all_patient_incidence_mat %>%
  column_to_rownames(var = "NCBI_SAMN") %>%
  as.matrix()
```


```{r, fig.show='hide'}
gheatmap(p =iqtree_p_sub1, data = all_patient_incidence_mat1,  colnames= FALSE,color = NULL)+ scale_fill_gradient(low = "white", high = "navy")  

```

> MSV pair pairwise vs overall SNV distribution distance histogram 
Figure 3a

```{r, fig.keep='first'}
# get the core genme distance matrix 
dist_mat <- dist.dna(as.DNAbin(dna_aln), model = "N", as.matrix = TRUE, pairwise.deletion = F)
#dist_mat <- dist.dna(as.DNAbin(dna_aln_no_outgroup), model = "N", as.matrix = TRUE, pairwise.deletion = F)
mst <- ape::mst(dist_mat)

#plot(mst)

# transpose matrix to long pairwise form
dist_mat_long <- t(combn(colnames(dist_mat),2))
dist_mat_long <- data.frame(dist_mat_long, dist = dist_mat[dist_mat_long])

# create transmission pair lists

# find all combinations
trans_pair <- MDHHS_test %>%
  ungroup()%>%
  select(NCBI_SAMN, source) %>%
  mutate(pair1 = source,
         pair2 = NCBI_SAMN) %>%
  select(-c(NCBI_SAMN, source))

trans_pair1 <- MDHHS_test %>%
  ungroup() %>%
  select(NCBI_SAMN, source) %>%
  mutate(pair1 = NCBI_SAMN,
         pair2 = source) %>%
  select(-c(NCBI_SAMN, source))

trans_pair <- rbind(trans_pair, trans_pair1)


# get isolate pairs that were putative transmission pairs
dist_mat_long$trans_pair <- 0
for( i in 1:nrow(trans_pair)){
  a <- trans_pair$pair1[i] # sample id in X1
  b <- trans_pair$pair2[i] # sample id in X2
  for(j in 1:nrow(dist_mat_long)){
    c <- dist_mat_long$X1[j]
    d <- dist_mat_long$X2[j]
 
  if( a==c & b==d){
    dist_mat_long$trans_pair[j] <- 1
  }
  if(a==d & b==c){
    dist_mat_long$trans_pair[j] <- 1
  }
    }
}


ggplot(dist_mat_long, aes(x = dist, fill = as.factor(trans_pair)))+geom_histogram(binwidth = 1) +labs(x = "SNV distance", y = "Number of pairs",fill = "MSV pair") + scale_fill_manual(values = c('0'='gray','1'='navy' ),labels = c("0"="No","1"="Yes")) + theme_bw() + theme(panel.grid = element_blank(),legend.position = "bottom")
#ggsave("/Users/tifwan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/SNV_threshold_compare.pdf",width = 6,height=6)
collection_HCF_df <- select(patient_loc_key_ID, Isolate_ID, HCF_exposure_1, Patient_ID)

collection_HCF_df <- merge(key, collection_HCF_df, by = "Isolate_ID")
collection_HCF_df <- select(collection_HCF_df, -Isolate_ID)
pt_NCBI_SAMN_key <- select(collection_HCF_df, Patient_ID, NCBI_SAMN) # key between NCBI_SAMN and Patient_ID
mst_g = graph.adjacency(mst, mode = "undirected")
gg = ggnetwork(mst_g, layout_with_fr(mst_g))
gg <- merge(gg, collection_HCF_df, by.x = "name",by.y = "NCBI_SAMN")

test <- gg
test$destination <- 0
for(i in 1:nrow(test)){
  for(j in 1:nrow(test)){
      if(test$x[i]==test$xend[j]& test$y[i]==test$yend[j]){
        test$destination[j] <- test$name[i]
      }
  }
}
test1 <- merge(test, pt_NCBI_SAMN_key, by.x = "destination", by.y = "NCBI_SAMN")
test1$trans_pair <- 0
for( i in 1:nrow(trans_pair)){
  a <- trans_pair$pair1[i] # sample id in X1
  b <- trans_pair$pair2[i] # sample id in X2
  for(j in 1:nrow(test)){
    c <- test1$name[j]
    d <- test1$destination[j]
 # whether these samples have been assigned as transmission pairs
  if(test1$Patient_ID.x[j]==test1$Patient_ID.y[j]){
    test1$trans_pair[j] <- 1
  }
    
  if( a==c & b==d){
    test1$trans_pair[j] <- 2
  }
  if(a==d & b==c){
    test1$trans_pair[j] <- 2
  }
    
    }
}

ggplot(test1, aes(x = x, y = y, xend = xend, yend = yend)) +
  geom_edges(aes(color = as.factor(trans_pair)), alpha = 0.5, curvature = 0.1) +
  geom_nodes(aes(color = HCF_exposure_1), size = 2) + theme_blank()  +theme(legend.position = "none")+
  scale_color_manual(values = location_col)

```
Find pairwise distance for genomic only pairs
```{r}
condition = "Genomic only"
dist_val = c()
df <- trans_edge_test3 %>%
  filter(CL==condition)
for (i in 1:nrow(df)){
  new_val <- dist_mat[df$from[i],df$to[i]]
  dist_val = c(dist_val, new_val)
}
dist_mat["SAMN24979406","SAMN24979412"]
dist_mat["SAMN21919088","SAMN18683849"]
dist_mat["SAMN16828094","SAMN17309430"]
dist_mat[	"SAMN16828097","SAMN17309427"]

hist(dist_val)
summary(dist_val)
genomic_only_df <- patient_loc_ID_long[patient_loc_ID_long$NCBI_SAMN%in%unique(df$to),] %>%
  filter(HCF!="")

```

# the pairwise distance among isolates from the same hospital
```{r}

patient_loc_ID_long_filter <- patient_loc_ID_long %>%
  filter(HCF!="")
pairwise_snv_epi <- data.frame(
  pt1 = c(),
  pt2 = c(),
  HCF = c(),
  SNV = c(),
  prediction = c()
)
patient_pair <- as.data.frame(t(combn(unique(trans_edge_test3$to),2)))
patient_pair_rev <- data.frame(
  V1 = patient_pair$V2,
  V2 = patient_pair$V1
)
patient_pair <- rbind(patient_pair, patient_pair_rev)
for (i in 1:nrow(patient_pair)){
  p1 <- patient_pair$V1[i]
  p2 <- patient_pair$V2[i]
  p1_loc <- patient_loc_ID_long_filter$HCF[patient_loc_ID_long_filter$NCBI_SAMN==p1]
  p2_loc <- patient_loc_ID_long_filter$HCF[patient_loc_ID_long_filter$NCBI_SAMN==p2]
  overlap <- intersect(p1_loc,p2_loc)
  if(!is.null(overlap)){
    for(hcf in overlap){
      snv <- dist_mat[p1,p2]
      if(length(trans_edge_test3$CL[trans_edge_test3$from==p1 & trans_edge_test3$to==p2])!=0){
        new_df <- data.frame(
        pt1 = p1,
        pt2 = p2,
        HCF = hcf,
        SNV = snv,
        prediction = trans_edge_test3$CL[trans_edge_test3$from==p1 & trans_edge_test3$to==p2]
      )
      }
      else if(length(trans_edge_test3$CL[grepl(trans_edge_test3$from,hcf) & trans_edge_test3$to==p1])!=0){
        new_df <- data.frame(
        pt1 = p1,
        pt2 = p2,
        HCF = hcf,
        SNV = snv,
        prediction = trans_edge_test3$CL[grepl(trans_edge_test3$from,hcf) & trans_edge_test3$to==p1]
      )
      }
      else{
        new_df <- data.frame(
        pt1 = p1,
        pt2 = p2,
        HCF = hcf,
        SNV = snv,
        prediction = "None"
        )
      }
      
      pairwise_snv_epi <- rbind(pairwise_snv_epi,new_df)
    }
  }
}

pairwise_snv_epi_cleaned <- pairwise_snv_epi %>%
  mutate(pair_id = pmap_chr(list(pt1, pt2), ~ paste(sort(c(..1, ..2)), collapse = "_"))) %>%
  distinct(pair_id, across(colnames(pairwise_snv_epi)[3:5]), .keep_all = TRUE) %>%
  select(-pair_id)%>%
  merge(patient_loc_ID_long[,c("variable","NCBI_SAMN","HCF","CollectionDate")]%>%filter(variable==1),by.x = c("pt1","HCF"),by.y = c("NCBI_SAMN","HCF")) %>%
  mutate(pt1_date = CollectionDate) %>%
  select(-c(CollectionDate,variable)) %>%
  merge(patient_loc_ID_long[,c("variable","NCBI_SAMN","HCF","CollectionDate")]%>%filter(variable==1),by.x = c("pt2","HCF"),by.y = c("NCBI_SAMN","HCF")) %>%
  mutate(pt2_date = CollectionDate) %>%
  select(-c(CollectionDate,variable))

pairwise_snv_epi_genomic_only <- pairwise_snv_epi_cleaned %>%
  filter(pt1 %in% unique(genomic_only_df$NCBI_SAMN) |pt2 %in%unique(genomic_only_df$NCBI_SAMN)) %>%
  mutate(case = ifelse(pt1%in%unique(genomic_only_df$NCBI_SAMN),"pt1","pt2"),
         epi = case_when(case=="pt1" & pt1_date>pt2_date~"missed",
                         case=="pt2" & pt2_date>pt1_date~"missed")) 


summary(pairwise_snv_epi_genomic_only$SNV[pairwise_snv_epi_genomic_only$epi=="missed"])
```



```{r}
## print values for manuscript
sorted_sample_collection_dates <- arrange(patient_loc_key_ID, Specimen_collection_date) %>%
  select(Specimen_collection_date) 
print("observed duration of outbreak ")
print(paste("First case:",sorted_sample_collection_dates$Specimen_collection_date[1]))
print(paste("Last case:",sorted_sample_collection_dates$Specimen_collection_date[length(sorted_sample_collection_dates$Specimen_collection_date)] ))
print("First isolates that are HCF acquired at the sample collection HCF")
patient_loc_key_ID %>%
  mutate(Specimen_collection_date = as.Date(Specimen_collection_date, format = "%m-%d-%y",origin = "1970-01-01"),
        HCF_1_admit_dt = as.Date(HCF_1_admit_dt, format = "%m/%d/%y",origin = "1970-01-01"),
         time_to_collection = Specimen_collection_date-HCF_1_admit_dt,
         HCF_acquired = ifelse(Specimen_collection_date-HCF_1_admit_dt>2,"Yes","No")) %>%
  distinct(Patient_ID, .keep_all = TRUE) %>%
  group_by(HCF_acquired) %>%
  summarise(N=n())


print(paste("All unique of HCF exp. among patients:",length(unique(HCF_info$HCF_name)),"HCFs"))
print(paste("All unique of sample collection HCF among patients:",length(unique(patient_loc_key_ID$HCF_exposure_1)),"HCFs"))
print(paste("Total number of isolates:",length(unique(patient_loc_ID_long1$NCBI_SAMN)),"isolates"))
print(paste("Total number of patients:",length(unique(patient_loc_ID_long1$Patient_ID)),"isolates"))
paste("counts for different HCF types")
table(HCF_info$HCF_type)
paste("Counts of isolate type")
table(patient_loc_key_ID$`Isolate Type`)

print("Breakdown of assigned epi_type by confidence level")
unique_transmissions <- trans_edge_test3  %>%
  ungroup() %>%
  group_by(HCF1,HCF2, pt2, CL) %>% # finding the same healthcare facility where NDM transmitted to the same pt2 
  arrange(pt1_date) %>% # sort by the dates of the source so the earliest one will come first
  mutate(shared_HCF = row_number(), 
         CL = ifelse(HCF1==HCF2,"intra",CL),
         to_delete = ifelse(shared_HCF!=1,1,0)) %>% ## only keeping first observed transmission event  
  filter(to_delete==0) %>% 
  ungroup()
unique_inter <- unique_transmissions %>%
  filter(HCF1!=HCF2)
table(unique_inter$HCF1)
table(unique_inter$HCF2)

print(paste("# Within-hospital outbreaks:",length(unique(trans_edge_test3$HCF2[trans_edge_test3$HCF1==trans_edge_test3$HCF2]))
))
length(unique(trans_edge_test3$pt2[trans_edge_test3$HCF1==trans_edge_test3$HCF2]))

# Find numbers of isolates in each group of specific CL epitypes
samples_w_high_CL <- unique(unique_transmissions$pt2[unique_transmissions$CL!="Uncertain"]) # isolates with specific source attributions
samples_w_genom_epi <- unique(unique_transmissions$pt2[unique_transmissions$CL=="Genomic & Epi"]) # isolates with Genomic & epi links
samples_w_intra <- unique(unique_transmissions$pt2[unique_transmissions$CL=="intra"]) # isolates with Genomic & epi links

samples_w_genomic_only <- unique(unique_transmissions$pt2[unique_transmissions$CL=="Genomic only"]) # isolates with genomic only sources
samples_w_low_CL <- unique(unique_transmissions$pt2[unique_transmissions$CL=="Uncertain"]) # isolate with non-specific source attribution
setdiff(samples_w_low_CL,samples_w_high_CL) ## find the isolates that only had uncertain source attributions
setdiff(samples_w_genomic_only, samples_w_intra)

print("Breakdown of epilinks in each group of specific CL")
print(paste("# High confidence source attribution:", length(samples_w_high_CL)))
print(paste("# Genomically and Epi:", length(setdiff(samples_w_genom_epi, samples_w_intra))))
print(paste("# Intra-facility:", length(samples_w_intra)))
print(paste("# Genomic only :", length(samples_w_genomic_only))) 
print(paste("# Uncertain:", length(setdiff(samples_w_low_CL,samples_w_high_CL))))

print("The cgSNVs summary")
summary(dist_mat_long$dist[dist_mat_long$trans_pair==1])
```


```{r}
load("/Users/tifwan/RA/MDHSS/jupyter_notebook/2023_01_16_flow_summary.RData")

```

```{r}
#duplicate the flow output data from MI_connect data
patient_flow_output1 <- patient_flow_output 

# extract the loc1 to loc2 df w/ metrics
patient_flow_output_12 <- patient_flow_output1 %>%
  select(loc1, loc2, n_transfers_f12, pt_trans_metric_f12) # keep loc1, loc2 n_transfers and pt_trans_metric

# extract the loc2 --> loc1 df w/ metrics 
patient_flow_output_21 <- patient_flow_output1 %>% # subset data to the data with locations and the metrics from loc2 to loc1
  select(loc1, loc2, n_transfers_f21, pt_trans_metric_f21) %>% 
  mutate(new_loc1 = loc2, # temporary name for loc1
         new_loc2 = loc1, # temporary name for loc2
         new_n_transfers_f12 = n_transfers_f21, # temporary name
         new_pt_trans_metric_f12 = pt_trans_metric_f21) %>% # temporary name
  select(-c(loc1, loc2, n_transfers_f21, pt_trans_metric_f21)) %>% # delete temporary vars
  mutate(loc1 = new_loc1, # rename 
         loc2 = new_loc2, # rename
         n_transfers_f12 = new_n_transfers_f12, # rename
         pt_trans_metric_f12 = new_pt_trans_metric_f12) %>% # rename
  select(-c(new_loc1, new_loc2, new_n_transfers_f12,new_pt_trans_metric_f12)) # delete temporary vars

# bind the rows from patient_flow_output_12  and patient_flow_output_21
patient_flow_output2 <- rbind(patient_flow_output_12, patient_flow_output_21)
```


```{r}
## create index for MDHHS and MI_conncet HCF names by CCN
MI_HCF_info <- data.frame(MI_HCF = c(MI_2021$facname1, MI_2021$facname2),
                          CCN = c(MI_2021$numid1, MI_2021$numid2)) 
MI_HCF_info <- MI_HCF_info[!duplicated(MI_HCF_info$MI_HCF),] # only keep unique obs
MI_HCF_info$CCN <- as.character(MI_HCF_info$CCN)
# finding corresponding names in MI connect with HCF_info from MDHHS
#patient_flow_output <- get_patient_flow(pt_flow2)
# merge CCN with HCF_info
MDHHS_CCN1 <- merge(MDHHS_CCN, MI_HCF_info, by.x = "HCF_CCN", by.y = "CCN", all.x = TRUE) 
```

```{r}

# add CCN to pt_flow output2
patient_flow_output3 <- merge(patient_flow_output2, MDHHS_CCN1, by.x = "loc1", by.y= "MI_HCF", all = TRUE ) # merge by loc1
patient_flow_output3$HCF_CCN1 <- patient_flow_output3$HCF_CCN # rename var
patient_flow_output3$HCF_name1 <- patient_flow_output3$HCF_name # rename var
patient_flow_output3 <- patient_flow_output3%>%
  select(-c(HCF_CCN, HCF_name))

patient_flow_output3 <- merge(patient_flow_output3, MDHHS_CCN1, by.x = "loc2", by.y= "MI_HCF", all = TRUE ) # merge by loc2
patient_flow_output3$HCF_CCN2 <- patient_flow_output3$HCF_CCN # rename var
patient_flow_output3$HCF_name2 <- patient_flow_output3$HCF_name # rename var
patient_flow_output3 <- patient_flow_output3%>%
  select(-c(HCF_CCN, HCF_name))

# reformat the transfer 
patient_flow_output4 <- patient_flow_output3 %>%
  mutate(loc1 = ifelse(!is.na(HCF_name1), HCF_name1, loc1),
         loc2 = ifelse(!is.na(HCF_name2), HCF_name2, loc2)) %>%
  select(-c(HCF_CCN1, HCF_CCN2, HCF_name1, HCF_name2))
```


```{r}
trans_count_sum <- trans_edge_test3 %>%
  distinct() %>% # make sure all transmission events were unique
  group_by(HCF1, HCF2, CL) %>%
  summarise(sum = n())

# create vector of all transfer location
HCF_v <- unique(c(trans_count_sum$HCF1, trans_count_sum$HCF2))

# only keep HCF transfers with both MDHHS HCF in either HCF1 or HCF2
patient_flow_output5 <- patient_flow_output4 %>%
  filter(loc1 %in% HCF_v | loc2 %in%HCF_v) %>%
  select(-n_transfers_f12) # delete n_transfers row

# add and prepare variables to correspond patient_flow_output5 with epi_transfer_df
patient_flow_output5 <- patient_flow_output5 %>%
  mutate(epi_type = 0, # add epi_type variable to correspond with epi_transfer_df
         trans_type = "MI_connect")
```


```{r}
trans_MDHHS <- trans_edge_test3 %>%
  distinct() %>%
  mutate(loc1 = HCF1, # rename HCF1 to loc1 to correspond to patient_flow_output5
         loc2 = HCF2, # rename HCF2 to loc2 to correspond to patient_flow_output5
         pt_trans_metric_f12 = 0,
         trans_type = "MDHHS",# add pt_trans_metric_f12 to correspond to patient_flow_output5
         epi_type = CL) %>% 
  ungroup() %>%
  select(loc1,loc2,pt_trans_metric_f12, epi_type, trans_type) %>%
  filter(loc1!=loc2)

# rbind 
pt_flow_epi <- rbind(trans_MDHHS , patient_flow_output5)

for(i in 1:nrow(pt_flow_epi)){
  trans_type_i <- pt_flow_epi$trans_type[i]
  
}
to_delete_i <- NULL
## populate the MDHHS rows with pt_trans_metrics if there is a value in MI_connect
for(i in which(pt_flow_epi$trans_type=="MDHHS")){ # i: position of MDHHS samples
  for(j in which(pt_flow_epi$trans_type=="MI_connect")){ # position of MI_connect samples
    if(pt_flow_epi$loc1[i]==pt_flow_epi$loc1[j] & pt_flow_epi$loc2[i]==pt_flow_epi$loc2[j]){ # for same in MDHHS and MI_connect transfers
      pt_flow_epi$pt_trans_metric_f12[i] <- pt_flow_epi$pt_trans_metric_f12[j] # assign the MI_connect pt_trans_metrics value to the MDHHS positions (i)
      to_delete_i <- c(to_delete_i, j) # delete the duplicated transfers shown in both MI_connect and MDHHS

    } 
  }
}

pt_flow_epi1 <- pt_flow_epi[-to_delete_i,] # delete values in vector to_delete
pt_flow_epi1 <- pt_flow_epi1%>%
  filter(loc1!=loc2) %>% # remove intra-facility transfers
  filter(pt_trans_metric_f12!=0) # remove transfer metrics values that are 0


# create data frame for each epi_type separately
trans_list <- vector(mode = 'list', length=3) # empty list
names(trans_list) <- unique(pt_flow_epi1$epi_type)[1:3] # names from pt_flow_epi1$epitype
for ( i in names(trans_list)){
  # 1. populate with MDHHS data first 
  # add the MDHHS samples to the respective dataframes for each epi_type
  trans_list[[i]] <- pt_flow_epi1[pt_flow_epi1$epi_type==i,] ## the rows that belonged to this particular epi_type
  index <- which (pt_flow_epi1$epi_type==i) 
  # 2. populate with MI_connect : the controls for each group
  for(j in which(pt_flow_epi1$trans_type=="MI_connect")) { # for MI_connect samples
    if (pt_flow_epi1$loc1[j]%in%pt_flow_epi1$loc1[index]){ # for those that has the same source HCF as MDHHS samples
      trans_list[[i]] <- rbind(trans_list[[i]], pt_flow_epi1[j,]) # add these rows to the trans_list
    }
  }
  trans_list[[i]]$epi_type <- i # variable names from pt_flow_epi1$epitype
}


# transmission df for plotting data
trans_df <- rbindlist(trans_list) # bind the list to a df
```

```{r}
# pt_flow_epi1 %>%
#   mutate(epi_type = ifelse(epi_type==0, "MI_connect",epi_type)) %>%
#   ggplot(aes(x = fct_relevel(epi_type, "MI_connect", "Genomic & Epi","Genomic only","Uncertain"), y = log(pt_trans_metric_f12), fill = trans_type, color = trans_type ))+ geom_violin(colour = NA, alpha = .6, draw_quantiles = .5) + scale_fill_manual (values = c("#F4C434","#4378A2"))+ geom_boxplot(color = "black", width = 0.2)

```

## Comparing the patient flow of MDHHS predictions and CMS transfer data
> Directed predicted transmission violin plot 
Figure 3c

```{r}
# violin plot
# ggplot(trans_df, aes(x = trans_type, y = log(pt_trans_metric_f12), fill = trans_type, color = trans_type )) + geom_violin(colour = NA, alpha = .6, draw_quantiles = .5) +
#   facet_wrap(.~epi_type, nrow = 1) + scale_fill_manual (values = c("#F4C434","#4378A2"))

# renaming values to correspond to the needed y -axis
trans_df %>%
  mutate(epi_type = ifelse(epi_type=="Uncertain","multiple possible\n source facility", epi_type)) %>%
ggplot(aes(x = trans_type, y = log(pt_trans_metric_f12), fill = trans_type, color = trans_type)) + geom_violin(colour = NA, alpha = .6, draw_quantiles = .5) +
  facet_wrap(.~epi_type, nrow = 1) + scale_fill_manual (name = "Predicted \ntransmission",values = c("#F4C434","#4378A2"), labels = c("MDHHS" = "Yes","MI_connect" ="No"))+ geom_boxplot(color = "black", width = 0.2) + labs(x = "Predicted transmission between facilities", y = "Inter-facility patient transfer") + scale_x_discrete(labels = c("MDHHS"="Yes", "MI_connect"= "No")) + theme_bw() + theme(panel.grid = element_blank())
#ggsave("/Users/tifwan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/directed_violin_predictions.pdf",width =7, height=4)
#ggsave("/Users/tiffanywan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/updated_figures2/directed_violin.pdf",width = 10, height = 7)#ggsave("/Users/tiffanywan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/updated_figures2/directed_violin.pdf",width = 10, height = 7)#ggsave("/Users/tiffanywan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/updated_figures2/directed_violin.pdf",width = 10, height = 7)
```


```{r}
trans_df1 <- trans_df %>%
filter(!is.na(pt_trans_metric_f12))
aggregate(trans_df1 %>% select(where(is.numeric)), list(trans_df1$epi_type, trans_df1$trans_type), mean)

# get the numbers of transfers in each epi_type
print("Genomic & Epi")
table(trans_df[trans_df$epi_type=="Genomic & Epi",]$trans_type)
print("Genomic only")
table(trans_df[trans_df$epi_type=="Genomic only",]$trans_type)
print("Uncertain")
table(trans_df[trans_df$epi_type=="Uncertain",]$trans_type)
```


```{r, fig.show='hide'}
## plot distribution of transfer metrics 
ggplot(trans_df, aes(x = pt_trans_metric_f12, fill = trans_type)) + geom_histogram(binwidth = 0.005) +
    facet_wrap( nrow = 2, epi_type ~ trans_type, scales = "free", dir = "v")

## Plot adjusted distribution
# ggplot(trans_df, aes(x = log(pt_trans_metric_f12), fill = trans_type)) + geom_histogram(binwidth = 0.2) +
#     facet_wrap( nrow = 2, epi_type ~ trans_type, scales = "free", dir = "v")

ggplot(trans_df, aes(x = log(pt_trans_metric_f12), fill = trans_type, color = trans_type)) + geom_density(alpha=.3) +
    facet_wrap( ncol =1,epi_type~., scales = "free_y")
```


```{r}
###############statistical tests STARTS ##################################################
# comparing MDHHS results with MI_connect results in each group
for(i in unique(trans_df$epi_type)){
  test <- wilcox.test(x = trans_df$pt_trans_metric_f12[trans_df$trans_type=="MDHHS" & trans_df$epi_type==i], y = trans_df$pt_trans_metric_f12[trans_df$trans_type=="MI_connect" & trans_df$epi_type==i], paired=FALSE,conf.int = TRUE)
  print(paste("The p-value between MDHHS and MI_connect data within transmission type", i, test$p.value))
}

trans_df_MDHHS <- trans_df %>%
  filter(trans_type=="MDHHS")

# fit linear model
trans_df <-trans_df %>%
  group_by(epi_type) %>%
  mutate(Confidence = case_when(trans_type=="MI_connect" ~ 1, ## turning confidence levels into numeric values
                                epi_type=="Genomic & Epi"~4, 
                                epi_type=="Genomic only" ~3,
                                epi_type=="Uncertain" ~2)) %>%
  filter(!is.na(pt_trans_metric_f12) & pt_trans_metric_f12!=0) 

fit <- lm(log(pt_trans_metric_f12) ~ as.factor(Confidence), data = trans_df) 
summary(fit)

# testing the difference between MI_connect/MDHHS within each transmission type using Wilcoxon Signed rank test
pair_tests <- lapply(trans_list, function(x){
  wilcox.test(pt_trans_metric_f12 ~ trans_type,data=x, conf.int = TRUE)
}
)

# kruskal.test(pt_trans_metric_f12~epi_type,
#              data = trans_df[trans_df$trans_type=="MDHHS"]) 
# 
# kruskal.test(pt_trans_metric_f12~epi_type,
#              data = trans_df[trans_df$trans_type=="MI_connect"]) 
##############################statistical test ENDS ##################################
```


```{r}
#### INDIRECT TRANSFERS STARTS ###################
# include every possible direction of transfers that may occur in the patient_transfer_output dataset
# create opposite directions for indirect links
epi_transfer_df_21 <-trans_MDHHS %>% # create new transfer data from the opposite direction of the transfers from MDHHS
  mutate(loc1_temp = loc2,
         loc2_temp = loc1) %>%
  select(-c(loc2, loc1)) %>%
  mutate(loc1 = loc1_temp,
         loc2 = loc2_temp) %>%
  select(-c(loc1_temp, loc2_temp))

epi_transfer_df_com <- rbind(trans_MDHHS,epi_transfer_df_21) # combine the transfers and opposites

#### Only Include transfers loc1 and loc2 that matches in the transfer_patient output data

# add CCN to pt_flow output2
patient_flow_output_test <- merge(patient_flow_output5, MDHHS_CCN1, by.x = "loc1", by.y= "MI_HCF", all = TRUE ) # merge by loc1
patient_flow_output_test$HCF_CCN1 <- patient_flow_output_test$HCF_CCN # rename var
patient_flow_output_test$HCF_name1 <- patient_flow_output_test$HCF_name # rename var
patient_flow_output_test <- patient_flow_output_test%>%
  select(-c(HCF_CCN, HCF_name))

patient_flow_output_test <- merge(patient_flow_output_test, MDHHS_CCN1, by.x = "loc2", by.y= "MI_HCF", all = TRUE ) # merge by loc1
patient_flow_output_test$HCF_CCN2 <- patient_flow_output_test$HCF_CCN # rename var
patient_flow_output_test$HCF_name2 <- patient_flow_output_test$HCF_name # rename var
patient_flow_output_test <- patient_flow_output_test%>%
  select(-c(HCF_CCN, HCF_name))


# reformat the transfer 
patient_flow_output_test2 <- patient_flow_output_test %>%
  mutate(loc1 = ifelse(!is.na(HCF_name1), HCF_name1, loc1),
         loc2 = ifelse(!is.na(HCF_name2), HCF_name2, loc2)) %>%
  select(-c(HCF_CCN1, HCF_CCN2, HCF_name1, HCF_name2))

pt_flow_filter <- patient_flow_output_test2 %>%
  filter(loc1 %in% HCF_v | loc2 %in%HCF_v) %>%
  mutate(trans_type = "MI_connect",
         epi_type = 0)

# epi_transfer_df_com <- epi_transfer_df_com %>%
#   mutate(trans_type = "MDHHS",
#          n_transfers_f12= 0,
#          pt_trans_metric_f12 = 0,
#          n_transfers_f21 = 0,
#          pt_trans_metric_f21 = 0,
#          sum_transfers = 0,
#          sum_pt_trans_metric = 0) %>%
#   select(-match)


 # put together MI_connect pt_flow output and  actual MDHHS transfers and delete repeated transfers
# rbind 
pt_flow_epi2 <- rbind(epi_transfer_df_com, pt_flow_filter)
pt_flow_epi2 <- pt_flow_epi2 %>%
  distinct() %>%
  filter(!is.na(pt_trans_metric_f12))

# delete obs that are duplicated in MDHHS and MI_connect
to_delete_i <- NULL
for(i in which(pt_flow_epi2$trans_type=="MDHHS")){
  for(j in which(pt_flow_epi2$trans_type=="MI_connect")){
    if(pt_flow_epi2$loc1[i]==pt_flow_epi2$loc1[j] & pt_flow_epi2$loc2[i]==pt_flow_epi2$loc2[j]){
      pt_flow_epi2$pt_trans_metric_f12[i] <- pt_flow_epi2$pt_trans_metric_f12[j] 
      to_delete_i <- c(to_delete_i, j)
    } 
  }
}
 
# 
# for(i in which(pt_flow_epi2$trans_type=="MDHHS")){
#   for(j in which(pt_flow_epi2$trans_type=="MI_connect")){
#     if(pt_flow_epi2$loc1[i]==pt_flow_epi2$loc1[j] & pt_flow_epi2$loc2[i]==pt_flow_epi2$loc2[j] & pt_flow_epi2$trans_type[j]=="MI_connect") # if MDHHS transfers shows up in the MI_connect as well, only show the MDHHS obs
#       to_delete_i <- c(to_delete_i, j)
#   }
# }

pt_flow_epi2  <- pt_flow_epi2 %>%
  filter(loc1!=loc2) %>% # remove intra-facility transfers
  filter(pt_trans_metric_f12!=0)

pt_flow_epi3 <- pt_flow_epi2[-to_delete_i,] # delete obs that are duplicated in MDHHS and MI_connect
pt_flow_epi3 <- pt_flow_epi3 %>%
  filter(!is.na(pt_trans_metric_f12))
colnames(pt_flow_epi3)
# ggplot(pt_flow_epi3, aes(x = trans_type, y = sum_pt_trans_metric)) + geom_violin() +
#   facet_grid(.~epi_type)


# create data frame for each epi_type separately
trans_list_indi <- vector(mode = 'list', length=3) # empty list
names(trans_list_indi) <- unique(pt_flow_epi3$epi_type)[1:3] # names from pt_flow_epi1$epitype
for ( i in names(trans_list_indi)){
  # 1. populate with MDHHS data first 
  # add the MDHHS samples to the respective dataframes for each epi_type
  trans_list_indi[[i]] <- pt_flow_epi3[pt_flow_epi3$epi_type==i,]
  index <- which (pt_flow_epi3$epi_type==i)
  # 2. populate with MI_connect : the controls for each group
  for(j in which(pt_flow_epi3$trans_type=="MI_connect")) {
    if (pt_flow_epi3$loc1[j]%in%pt_flow_epi3$loc1[index]){
      trans_list_indi[[i]] <- rbind(trans_list_indi[[i]], pt_flow_epi3[j,])
    }
  }
  trans_list_indi[[i]]$epi_type <- i # variable names from pt_flow_epi1$epitype
}

trans_df_indi <- rbindlist(trans_list_indi)


# trans_df_indi <- trans_df_indi %>%
#     mutate(epi_type = case_when(epi_type ==2~"2 inter-facility (acquirer)",
#                                 epi_type ==3~"3 inter-facility (source)",
#                                 epi_type ==4~"4 inter-facility (shared)",
#                                 epi_type ==5~"5 inter-facility (unknown)"))
```

> Undirected predicted transmission violon plot
Supplemental Figure 4
```{r,fig.keep="first"}
# violin plot
ggplot(trans_df_indi, aes(x = trans_type, y = log(pt_trans_metric_f12), fill = trans_type, color = trans_type)) + geom_violin(colour = NA, alpha = .6, draw_quantiles = .5) + 
  facet_wrap(.~epi_type, nrow = 1) + scale_fill_manual (values = c("#F4C434","#4378A2"))+ geom_boxplot(color = "black", width = 0.2) + labs(x = "Predicted transmission between facilities", y = "Inter-facility patient transfer") + scale_x_discrete(labels = c("MDHHS"="Yes", "MI_connect"= "No")) + theme_bw() + theme(panel.grid = element_blank())
ggsave("/Users/tifwan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/undirected_violin.pdf",width =7, height=4)
ggplot(trans_df_indi, aes(x = log(pt_trans_metric_f12), fill = trans_type, color = trans_type)) + geom_density(alpha=.3) +
    facet_wrap( ncol =1,epi_type~., scales = "free_y")
```

```{r}
# pt_flow_epi3 %>%
#   mutate(epi_type = ifelse(epi_type==0, "MI_connect",epi_type)) %>%
#   ggplot(aes(x = fct_relevel(epi_type, "MI_connect", "Genomic & Epi","Genomic only","Uncertain"), y = log(pt_trans_metric_f12), fill = trans_type, color = trans_type ))+ geom_violin(colour = NA, alpha = .6, draw_quantiles = .5) + scale_fill_manual (values = c("#F4C434","#4378A2"))+ geom_boxplot(color = "black", width = 0.2)

```

```{r}
########## statistical test STARTS here #######################

for(i in unique(trans_df_indi$epi_type)){
  test <- wilcox.test(x = trans_df_indi$pt_trans_metric_f12[trans_df_indi$trans_type=="MDHHS" & trans_df_indi$epi_type==i], y = trans_df_indi$sum_pt_trans_metric[trans_df_indi$trans_type=="MI_connect" & trans_df_indi$epi_type==i], paired=FALSE)
  print(paste("The Wilcoxin test p-value between MDHHS and MI_connect data within transmission type", i, test$p.value))
}

aggregate(trans_df_indi %>% select(where(is.numeric)), list(trans_df_indi$epi_type, trans_df_indi$trans_type), mean)


trans_df_indi_MDHHS <- trans_df_indi %>%
  filter(trans_type=="MDHHS")

# fit linear model
trans_df_indi <- trans_df_indi %>%
  group_by(epi_type) %>%
  mutate(Confidence = case_when(trans_type=="MI_connect" ~ 1, ## turning confidence levels into numeric values
                                epi_type=="Genomic & Epi"~4, 
                                epi_type=="Genomic only" ~3,
                                epi_type=="Uncertain" ~2)) %>%
  filter(!is.na(pt_trans_metric_f12) & pt_trans_metric_f12!=0) 

######### statistical test ENDS here ##########################
```

```{r load library}
#install.packages("graph4lg")
library(ggtree)
library(tidyverse)
library(readxl)
library(data.table)
library(aplot)
library(graph4lg)
library(ape)
library(phytools)
library(pheatmap)
library(BactDating)
library(phylobase)
library(TransPhylo)
library(lubridate)
library(coda)
library(mcmcplots)
library(seqinr)

```

```{r load 2023_02_03_Bactdaing_pwd.RData workspace for }
load("/Users/tifwan/RA/MDHSS/R_script/2023_02_03_Bactdating_pwd.RData")
```

```{r}
location_col <- c( "#800000",
                   "#de5d83",
                   "#872657",
                   "#cc6666",
                   "#cd5b45",
                   "#cc5500",
                   "#ffa700",
                   "#e48400",
                   "#cd7f32",
                   "#ffd700",
                   "#483d8b",
                   "#e5e500",
                   "#f4c2c2",
                   "#ffa6c9",
                   "#00ffff",
                   "#98777b",
                   "#996666",
                   "#fbceb1",
                   "#6f4e37",
                   "#808000",
                   "#702963",
                   "#00bfff",
                   "#915c83",
                   "#ff55a3",
                   "#704241",
                   "#bdb76b",
                   "#556b2f",
                   "#5f9ea0",
                   "#008000",
                   "#ccff00",
                   "#e0ffff",
                   "#ace1af",
                   "#7fffd4",
                   "#ccccff",
                   "#614051",
                   "#8db600",
                   "#a4c639",
                   "#cd9575",
                   "#008b8b",
                   "#89cff0",
                   "#195905",
                   "#004953",
                   "#9966cc",
                   "#e9d66b",
                   "#5d8aa8",
                   "#1560bd",
                   "#915f6d",
                   "gray50")

all_HCF_v <- c("Ascension Macomb Oakland - Madison Heights Campus",
               "Ascension Macomb Oakland - Warren Campus",
               "Ascension Providence Novi",
               "Ascension Providence Park Southfield",
               "Ascension St. John",
               "Beaumont Dearborn",
               "Beaumont Grosse Pointe",
               "Beaumont Royal Oak",
               "Beaumont Taylor",
               "DMC Detroit Receiving Hospital",
               "DMC Harper University Hospital",
               "DMC Sinai Grace Hospital",
               "Henry Ford Hospital",
               "Henry Ford Macomb Hospital",
               "Henry Ford Wyandotte Hospital",
               "Lake Huron Medical Center",
               "McLaren Port Huron",
               "Michigan Medicine",
               "Mid-Michigan Hospital West Branch",
               "Spectrum Butterworth Hospital",
               "John D. Dingell Department of Veterans Affairs Medical Center",
               "Karmanos Cancer Center",
               "Divinity Home Care",
               "DMC Rehabilitation Institute of Michigan",
               "Select Specialty Hospital Ann Arbor",
               "Select Specialty Hospital Downriver",
               "Select Specialty Hospital Grosse Pointe",
               "Select Specialty Hospital Northwest Detroit",
               "Vibra DMC Campus",
               "Vibra Taylor",
               "Advantage Living Center Northwest",
               "Advantage Living Center Samaritan",
               "Ambassador, A Villa Place",
               "Father Murray, A Villa Center",
               "Hamilton Nursing & Rehabilitation",
               "Hartford Nursing & Rehab",
               "Heartland Healthcare Center Grosse Pointe Woods",
               "Mission Point Nursing & Physical Rehab Center of Detroit",
               "Omni Continuing Care",
               "Regency at Bluffs Park",
               "Regency at Chene",
               "Regency on the Lake",
               "Riverview Health & Rehab Center Jefferson",
               "Riverview Health & Rehab Center North",
               "St. Joseph's Healthcare Center",
               "Villa at Rose City",
               "Wellbridge Rochester Hills")

# location_col <- c("#ffc0cb", "#6495ed", "#dda0dd", "#afeeee", "#ee82ee", "#98fb98", "#7fffd4", "#ffdead", "#ff69b4", "#2f4f4f", "#556b2f", "#8b4513", "#2e8b57", "#191970", "#8b0000", "#808000", "#008000", "#008080", "#b8860b", "#bdb76b", "#4682b4", "#d2691e", "#9acd32", "#00008b", "#32cd32", "#7f007f", "#8fbc8f", "#b03060", "#ffa500", "#ffa500", "#ffd700", "#ffff00", "#0000cd", "#00ff00", "#9400d3", "#00fa9a", "#dc143c", "#00ffff", "#00bfff", "#f4a460", "#9370db", "#f08080", "#adff2f", "#ff6347", "#b0c4de",  "red","black")
#unique(patient_loc_ID_long1$HCF)
location_col <- setNames(location_col , c(all_HCF_v,"Others"))
```


```{r read in tree and fasta file of variant calling results}
tree <- read.tree("/Users/tifwan/Desktop/flux_mount_gl/Project_MDHHS_genomics/old_Sequence_data/ST219_Public_genomes/2022_08_24_MDHHS_Public_variant_calling/2022_08_25_16_34_37_core_results/gubbins/iqtree_results/2022_08_25_16_34_37_SAMN26729713_genome_aln_w_alt_allele_unmapped.treefile")
fasta <- read.dna("/Users/tifwan/Desktop/flux_mount_gl/Project_MDHHS_genomics/old_Sequence_data/ST219_Public_genomes/2022_08_24_MDHHS_Public_variant_calling/2022_08_25_16_34_37_core_results/gubbins/2022_08_25_16_34_37_SAMN26729713_genome_aln_w_alt_allele_unmapped.filtered_polymorphic_sites.fasta",format = "fasta")
```


```{r}
setwd("~/RA/MDHSS/R_script")
patient_loc <- read.csv("../data/2022_08_25_location_time.csv")
patient_ID <- read_xlsx("/Users/tifwan/Dropbox (University of Michigan)/Michigan_CRE_Genomics/NDM Kpneumoniae ST219 7-11-22_update 10-4.xlsx",sheet = 2)
MDHHS_list <- fread("../data/MI_isolate_accn.tsv",)
iqtree <- read.tree("/Users/tifwan/Desktop/flux_mount_gl/Project_MDHHS_genomics/old_Sequence_data/ST219_Public_genomes/2022_08_24_MDHHS_Public_variant_calling/2022_08_25_16_34_37_core_results/gubbins/iqtree_results/2022_08_25_16_34_37_SAMN26729713_genome_aln_w_alt_allele_unmapped.treefile")
loc_date <- read_xlsx("../data/2022_08_25_location_time.xlsx")

# Isolate_ID, NCBI_SAMN
key <- fread("/Users/tifwan/Dropbox (University of Michigan)/Michigan_CRE_Genomics/Genome meta-data/st219_lookup_table.txt")
# fix the isolate ID and corresponding NCBI
```


```{r }
# drop tip SAMN26729713_2 
tree <- drop.tip(tree, "SAMN26729713_2")

###### compare variant calling pairwise distance with cophenenic pairwise distance ##############
#labels(fasta)
fasta_mat <- cbind(fasta)

fasta <- fasta[-which(labels(fasta)=="SAMN26729713_2"),]
# get pairwise distance: pairwise deletion==True 
dist_T <- dist.dna(fasta, model = "N", as.matrix = TRUE, pairwise.deletion = T) # non-core distances
dist_T_t <- t(combn(colnames(dist_T ),2))
dist_T_t <- data.frame(dist_T_t, dist = dist_T[dist_T_t])
# dist_T_t_MDHHS <- dist_T_t %>%
#   filter(X1%in%isolate_pt_key$NCBI_SAMN & X2 %in%isolate_pt_key$NCBI_SAMN)

dist_T_t <- dist_T_t %>%
  mutate(sample_type = case_when(X1%in%isolate_pt_key$NCBI_SAMN & X2 %in% isolate_pt_key$NCBI_SAMN ~ "MDHHS-MDHHS",
                                 X1%in%isolate_pt_key$NCBI_SAMN | X2 %in%isolate_pt_key$NCBI_SAMN ~ "MDHHS-Public",
                                 TRUE ~ "Public-Public"))
dist_T_MDHHS_t <- dist_T_t %>%
  filter(sample_type=="MDHHS-MDHHS")
```



```{r plot pairwise distance non core, fig.show='hide'}

ggplot(dist_T_t, aes(x= dist, fill = sample_type)) + geom_histogram(position = "identity" ,binwidth = 2, alpha = 0.5) +
  labs(x = "SNV", y = "# Pairwise distance", fill = "pair type") + scale_fill_viridis_d()

#ggsave("/Users/tifwan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/exploratory/var_dist_hist_MDHHS_public.pdf")
ggplot(dist_T_MDHHS_t, aes(x= dist)) + geom_histogram(position = "identity" ,binwidth = 1, alpha = 0.5) +
  labs() + scale_fill_viridis_d()
#ggsave("/Users/tifwan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/exploratory/var_dist_hist_MDHHS.pdf")
```



```{r pairwise distance core}
# get pairwise distance and transpose data set
dist_F <- dist.dna(fasta, model = "N", as.matrix = TRUE, pairwise.deletion = F) # core distances
dist_F_t <- t(combn(colnames(dist_F ),2))
dist_F_t <- data.frame(dist_F_t, dist = dist_F[dist_F_t])

dist_F_t <- dist_F_t%>%
  mutate(sample_type = case_when(X1%in%isolate_pt_key$NCBI_SAMN & X2 %in% isolate_pt_key$NCBI_SAMN ~ "MDHHS-MDHHS",
                                 X1%in%isolate_pt_key$NCBI_SAMN | X2 %in%isolate_pt_key$NCBI_SAMN ~ "MDHHS-Public",
                                 TRUE ~ "Public-Public"))
# only keeping MDHHS samples
dist_F_MDHHS_t <- dist_F_t %>%
  filter(sample_type == "MDHHS-MDHHS")
```

> Pairwise distance histogram
Supllemental Figure 2
```{r plot pairwise distance core,fig.keep="first"}
ggplot(dist_F_t, aes(x= dist, fill = sample_type)) + geom_histogram(position = "identity" ,binwidth = 2, alpha = 0.5) + scale_fill_viridis_d()+theme_bw() + theme(panel.grid = element_blank()) + labs(x = "SNV distance",y = "Number of pairs")
#ggsave("/Users/tifwan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/cg_dist_hist_MDHHS_public.pdf", height=4,width =7)
ggplot(dist_F_MDHHS_t, aes(x= dist)) + geom_histogram(position = "identity" ,binwidth = 1, alpha = 0.5) +
  labs(x = "SNV distance",y = "Number of pairs") + scale_fill_viridis_d() +theme_bw() + theme(panel.grid = element_blank())
#ggsave("/Users/tiffanywan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/exploratory/cg_dist_hist_MDHHS.pdf")
```


```{r,fig.show='hide'}
### get pairwise distance of tips usings cophenetic (from ape)
cophen <- cophenetic(tree)
cophen_t <- t(combn(colnames(cophen),2))
cophen_t <- data.frame(cophen_t, dist = cophen[cophen_t])

cophen_t <- cophen_t %>%
  mutate(sample_type = case_when(X1%in%isolate_pt_key$NCBI_SAMN & X2 %in% isolate_pt_key$NCBI_SAMN ~ "MDHHS-MDHHS",
                                 X1%in%isolate_pt_key$NCBI_SAMN | X2 %in%isolate_pt_key$NCBI_SAMN ~ "MDHHS-Public",
                                 TRUE ~ "Public-Public"))

cophen_MDHHS_t <- cophen_t %>%
  filter(sample_type =="MDHHS-MDHHS")

ggplot(cophen_t, aes(x= (median(dist_T_t$dist)/
median(dist))*dist, fill = sample_type)) + geom_histogram(position = "identity" ,binwidth = 2, alpha = 0.5) +
  labs(title = "cophenetic pairwise distance") + scale_fill_viridis_d()

median(dist_T_t$dist)
median(cophen_t$dist)

ggplot(cophen_MDHHS_t, aes(x= (median(dist_T_MDHHS_t$dist)/
median(dist))*dist)) + geom_histogram(position = "identity" ,binwidth = 1, alpha = 0.5) +
  labs(title = "cophenetic MDHHS pairwise distance") + scale_fill_viridis_d()
```

> Check convergence of Bactdating trees

```{r, results=TRUE}
# plot trace plots to check convergence
for (i in 1:length(time_tree)){
  print(paste("Model:", names(time_tree[i])))
  plot(time_tree[[i]],"trace")
  title(main = paste("Model:", names(time_tree[i])), line = 5)
}

```
```{r}
model_info <- t(sapply(time_tree, function(i){
  mcmc1=as.mcmc.resBactDating(i)
  c(effectiveSize(mcmc1), summary(mcmc1)$statistics[1,1], i$rootdate, i$CI[which.min(rowSums(i$CI)),], i$dic)
}))

colnames(model_info) <- c(' ESS',' ESS',' ESS','Clock rate','Root date','Root date lower CI','Root date upper CI','DIC')

summary_model <- as_tibble(round(model_info, 2)) %>% mutate(Model = rownames(model_info), .before=1)
summary_model
#write_csv(summary_model,"~/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/bactdating_model_stats.csv")
```
> Plot Epicurve
Figure 1a

```{r}

key[Isolate_ID=="20MP010425","NCBI_SAMN"] <- "20MP010425-WI-M3453-201110"
key[Isolate_ID=="20MP010444","NCBI_SAMN"] <- "20MP010444-WI-M5194-201116_S17"

# get Isolate_ID and Patient_ID
key2 <- patient_ID %>%
  select(Isolate_ID, Patient_ID)
key3 <- merge(key, key2, by = "Isolate_ID")

# renaming observations loc_date
patient_loc <- apply(patient_loc,2,function(x){gsub('Henry Ford Macomb Hospital Hospital','Henry Ford Macomb Hospital',x)})
patient_loc <- apply(patient_loc,2,function(x){gsub('DMC Rehabillitation Institute of Michigan','DMC Rehabilitation Institute of Michigan',x)})
patient_loc <- apply(patient_loc,2,function(x){gsub('Regency at Bluffs','Regency at Bluffs Park',x)})
patient_loc <- apply(patient_loc,2,function(x){gsub('Regency at Bluffs Park Park','Regency at Bluffs Park',x)})
patient_loc <- apply(patient_loc,2,function(x){gsub('Ascension Providence Southfield','Ascension Providence Park Southfield',x)})
patient_loc <- as.data.frame(patient_loc)

# merge key with location data
patient_loc_key <- merge(key3, patient_loc,  by = "Patient_ID", all = TRUE) 
patient_loc_key <- patient_loc_key %>%
  mutate(Isolate_ID = Isolate_ID.x) %>%
  select(-c(Isolate_ID.x, Isolate_ID.y))


# merge location and isolate collection dates
patient_loc_key_ID <- merge(patient_ID, patient_loc_key, by = "Isolate_ID", all = TRUE)
patient_loc_key_ID <- as.data.table(patient_loc_key_ID) # change df to data.table before tranposing to long 

patient_loc_key_ID <- patient_loc_key_ID %>%
  mutate(Patient_ID = Patient_ID.x) %>%
  select(-c(Patient_ID.x,Patient_ID.y))

patient_loc_key_ID <- arrange(patient_loc_key_ID, Specimen_collection_date)

# read in metadata of patient transfer network
HCF_info <- fread("/Users/tifwan/RA/MDHSS/data/HCF_info_coord.csv") %>%
  select(-V1)
HCF_info$HCF_CCN <- gsub("A","",HCF_info$HCF_CCN)

# rename HCF_name variables to match the patient HCF transfer network
HCF_info$HCF_name<- gsub("Advantage Living Center Samaritan.*","Advantage Living Center Samaritan",HCF_info$HCF_name)
HCF_info$HCF_name<- gsub("Advantage Living Center Northwest.*","Advantage Living Center Northwest",HCF_info$HCF_name)
HCF_info$HCF_name<- gsub("Advantage Living Center Northwest.*","Advantage Living Center Northwest",HCF_info$HCF_name)
HCF_info$HCF_name<- gsub("Heartland Healthcare Center Grosse Pointe Woods.*","Heartland Healthcare Center Grosse Pointe Woods",HCF_info$HCF_name)
HCF_info$HCF_name<- gsub("Divinity Home Care.*","Divinity Home Care",HCF_info$HCF_name)
HCF_info$HCF_name<- gsub("DMC Rehabillitation Institute of Michigan","DMC Rehabilitation Institute of Michigan",HCF_info$HCF_name)
HCF_info$HCF_name<- gsub("Ascension Providence Southfield","Ascension Providence Park Southfield",HCF_info$HCF_name)

HCF_info$HCF_name <- trimws(HCF_info$HCF_name)
```


```{r}
#######################
###   epicurve      ###
#######################
# Keep variable: sample Sample ID, sample collection HCf, sample collection date and Patient ID
iso_location0 <- patient_loc_key_ID %>%
  select(NCBI_SAMN, HCF_exposure_1, Specimen_collection_date, Patient_ID)

# # filter data to keep one sample per patient
# iso_location1 <- iso_location0 %>%
#   group_by(Patient_ID)%>%
#   arrange(Specimen_collection_date)%>%
#   mutate(row_num = row_number())%>%
#   filter(row_num==1) # only keep one isolate per patient
# 
# 
# # rename missing HCF type HCF
# iso_location[which(is.na(iso_location$HCF_type)),"HCF_type"] <- "Veteran Affairs Hospital" 
# 
# # get the unique hospitals
# iso_loc_sum <- iso_location %>%
#   group_by(HCF_exposure_1, HCF_type) %>%
#   summarise(n = n())
# table(iso_location$HCF_type)
# 
# # create the anonymous labels
# iso_loc_sum <- iso_loc_sum %>%
#   group_by(HCF_type) %>%
#   mutate(num = row_number(), # the numerator
#          new_names = paste(HCF_type,num)) # the labels
# 
# # merge back again the iso_location  
# iso_location1 <- merge(iso_location0, iso_loc_sum, by = "HCF_exposure_1")
iso_location1 <- merge(iso_location0, location_col_df1, by.x = "HCF_exposure_1",by.y = "HCF_name", all.x = TRUE)
# 
# legend_color <- merge(location_col_df, iso_location1, by.x = "HCF_name", by.y= "HCF_exposure_1", all = TRUE)
# legend_color <- merge(legend_color, HCF_info, by.x = "HCF_name",by.y = "HCF_name")
# legend_color <- legend_color %>%
#   arrange(HCF_type.y,  is.na(num),num) %>%
#   distinct(HCF_name,.keep_all = TRUE) %>%
#   group_by(HCF_type.y) %>%
#   mutate(num2 = row_number()) %>%
#   mutate(new_name2 = paste(HCF_type.y,num2))
# 
# HCF_pseodoname <- setNames(legend_color$new_name2, legend_color$color)
# #HCF_pseodoname <- order(HCF_pseodoname)
# ggplot(location_col_df1) + geom_bar(aes(pseudo_HCF, fill = color)) + scale_fill_identity("Healthcare facility",guide = "legend", labels = setNames(location_col_df1$pseudo_HCF,location_col_df1$color)) + theme(legend.position = "bottom") + guides(fill = guide_legend(ncol = 8)) + theme_void()
# 
# 
# # set location_col vector to data frame
# location_col_df <- data.frame(HCF_name = names(location_col),
#                               color = location_col)
# 
# iso_location1 <- merge(iso_location1, location_col_df1, by.x = "HCF_exposure_1", by.y = "HCF_name", all.x = TRUE)
# 
# iso_location1 <- iso_location1 %>%
#   mutate(color=ifelse(new_names=="Others", "gray50",color))
# 
# ## subset location data to new_names column and the corresponding colors
# new_names_HCF_df <- iso_location1 %>%
#   select(new_names,color) %>%
#   distinct()
```

```{r}
## set new_names color scheme
new_names_col <- c(new_names_HCF_df$color, "#0000FF",
                   "#008000",
                   "#FFFF00",
                   "#800080",
                   "#FFA500",
                   "#FFC0CB",
                   "#40E0D0",
                   "#A52A2A",
                   "#008080",
                   "#E6E6FA",
                   "#4B0082",
                   "#800000",
                   "#00FFFF",
                   "#808000",
                   "#FF00FF",
                   "#F5F5DC",
                   "#FFD700",
                   "#C0C0C0",
                   "#FFDAB9",
                   "#DDA0DD",
                   "#708090",
                   "#FF6B6B",
                   "#cc5500",
                   "black"
                   )
names(new_names_col) <- c(new_names_HCF_df$new_names,"NDM115",
               "NDM117",
               "NDM121",
               "NDM122",
               "NDM126",
               "NDM135",
               "NDM142",
               "NDM148",
               "NDM186",
               "NDM52",
               "NDM55",
               "NDM63",
               "NDM65",
               "NDM69",
               "NDM71",
               "NDM73",
               "NDM74",
               "NDM75",
               "NDM79",
               "NDM80",
               "NDM82",
               "NDM84",
               "NDM88",
               "none")
new_names_col

```


```{r, fig.keep = "first"}
# change variable format to date
iso_location1$Specimen_collection_date <- as.Date(iso_location1$Specimen_collection_date)
iso_location1 <- iso_location1 %>%
  group_by(HCF_exposure_1) %>%
  mutate(HCF_gt_3 = ifelse(n()<3, "Others",pseudo_HCF))
# plot epicurve
fig_1a <- ggplot(iso_location1, aes(x = Specimen_collection_date, fill = HCF_gt_3))+ geom_histogram(position="stack") + labs(x = "Specimen Collection Date", y= "Number of cases", fill = "Healthcare facility") +
  scale_x_date(date_breaks = "6 months",date_labels = "%b %Y ") + scale_fill_manual(values = c(setNames(location_col_df1$color,location_col_df1$pseudo_HCF),"Others"="gray50"))  + theme_bw() + theme(panel.grid.major = element_blank(),panel.grid = element_blank(), legend.position = "none")
fig_1a 
#ggsave("~/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/epicurve.pdf", width =8, height = 6)
epicurve_p <- iso_location1 %>%
  filter(HCF_gt_3!="Others") %>%
  ggplot(aes(x = Specimen_collection_date, fill = pseudo_HCF))+ geom_histogram(position="stack") + labs(x = "Specimen Collection Date", y= "Number of cases", fill = "Healthcare facility") +
  scale_x_date(date_breaks = "6 months",date_labels = "%b %Y ") + scale_fill_manual(values = c(setNames(location_col_df1$color,location_col_df1$pseudo_HCF),"Others"="gray50")) +theme(legend.position = "bottom", legend.key.spacing  = unit(0.5, "mm")) + theme_void()
fig_1_legend <- get_legend(epicurve_p)
#ggdraw(plot_grid(fig_1a,fig_1b,fig_1_legend,ncol = 3, rel_widths = c(3.5,3.5,1)))
ggsave("/Users/tifwan/University of Michigan Dropbox/Tiffany Wan/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/nature_communication_submission/manuscript_figures/2025_01_30_figures/fig_1.pdf", width = 20, height = 8)
 #ggsave("/Users/tifwan/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/epicurve_legend_bottom.pdf", width =10, height = 4)
```



> Plot time phylogeny using Bactdating
Figure 1b

```{r}
# save model output as phylo object
ggtree_ls <- lapply(time_tree, function(x){
  return(as.treedata.resBactDating(x))
})

HCF_alias <- iso_location1 %>%
  ungroup() %>%
  group_by(Patient_ID) %>%
  mutate(count = n(),
         id = ifelse(count==1,"none",Patient_ID)) %>%
  ungroup() %>%
  select(NCBI_SAMN, pseudo_HCF, id)

## Change the tip label colors if they are from the same patient
HCF_alias1 <- HCF_alias[match(ggtree_ls[["poisson"]][[1]][["tip.label"]], HCF_alias$NCBI_SAMN),]

# lapply(ggtree_ls,function(x){
#   p <- ggtree(x[[1]], mrsd="2022-05-27", # set start date
#               aes(col = c(HCF_alias1$id, rep("none",103)))) + # add branch colors
#     theme_tree2() 
#   x[[2]]$node <- as.numeric(x[[2]]$node)
#   p <- p %<+% x[[2]] + geom_range("length_0.95_HPD", color='gray50', size=2, alpha=.4) ## add the confidence level
#   p <- p %<+% HCF_alias + 
#  geom_tippoint(aes(color = new_names), alpha = 0.7) +
#    scale_color_manual(values = new_names_col) +
#   labs(color = "Healthcare facility") + geom_tiplab(aes(label=NA, col=id)) +
#     theme(legend.position = "none")
# 
# })
# 


p <- ggtree(ggtree_ls[["mixedcarc"]][[1]], mrsd="2022-05-27", # set start date
              aes(col = c(HCF_alias1$id, rep("none",103)))) + # add branch colors
    theme_tree2() 
ggtree_ls[["mixedcarc"]][[2]]$node <- as.numeric(ggtree_ls[["mixedcarc"]][[2]]$node)
p <- p %<+% ggtree_ls[["mixedcarc"]][[2]] + geom_range("length_0.95_HPD", color='gray50', size=2, alpha=.4) ## add the confidence level
p <- p %<+% HCF_alias + 
 geom_tippoint(aes(color = pseudo_HCF), alpha = 0.7) +
  scale_color_manual(values = c(setNames(location_col_df1$color,location_col_df1$pseudo_HCF),"Others"="gray50")) +
  labs(title = "mixedcarc",color = "Healthcare facility") + geom_tiplab(aes(label=NA, col=id)) +
    theme(legend.position = "none")
p
fig_1b <- p
  #ggsave("~/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/figures_v2/time_tree.pdf")
```
```{r}
p <- ggtree(ggtree_ls[["arc"]][[1]], mrsd="2022-05-27", # set start date
              aes(col = c(HCF_alias1$id, rep("none",103)))) + # add branch colors
    theme_tree2() 
ggtree_ls[["arc"]][[2]]$node <- as.numeric(ggtree_ls[["arc"]][[2]]$node)
p <- p %<+% ggtree_ls[["arc"]][[2]] + geom_range("length_0.95_HPD", color='gray50', size=2, alpha=.4) ## add the confidence level
p <- p %<+% HCF_alias + 
 geom_tippoint(aes(color = pseudo_HCF), alpha = 0.7) +
   scale_color_manual(values = c(setNames(location_col_df1$color,location_col_df1$pseudo_HCF),"Others"="gray50")) +
  labs(title = "arc",color = "Healthcare facility") + geom_tiplab(aes(label=NA, col=id)) +
    theme(legend.position = "none")
p

#ggsave("~/Dropbox (University of Michigan)/MDHHS regional genomics project/Manuscripts/2022 - ST219 regional transmission/figures/illustrator/bactdatingtree.pdf")
```

```{r}
## supplemental table
patient_metadata <- patient_loc_ID_long %>%
  select(NCBI_SAMN, HCF, Patient_ID, AdmitDate, DischargeDate, CollectionDate) %>%
  mutate(NCBI_SAMN = ifelse(is.na(CollectionDate), NA, NCBI_SAMN)) %>%
  merge(location_col_df1[,c("HCF_name","pseudo_HCF")], by.x = "HCF", by.y = "HCF_name") %>%
  mutate(Facility_ID = pseudo_HCF) %>%
  select(-c(HCF, pseudo_HCF)) 

patient_metadata$AdmitDate <- as.numeric(patient_metadata$AdmitDate)
patient_metadata$DischargeDate <- as.numeric(patient_metadata$DischargeDate)
patient_metadata$CollectionDate <- as.numeric(patient_metadata$CollectionDate)


patient_metadata$AdmitDate <- patient_metadata$AdmitDate - 18185
patient_metadata$DischargeDate <- patient_metadata$DischargeDate - 18185
patient_metadata$CollectionDate <- patient_metadata$CollectionDate - 18185
library(writexl)
write_xlsx(patient_metadata, "~/Dropbox (University of Michigan)/Michigan_CRE_Genomics/Patient_metadata/supplement_table.xlsx")
```

